<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Caitlin Page" />

<meta name="date" content="2024-10-25" />

<title>00_differentialMethylationSequencing</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">2_ontology_bias</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/caitlinpage/gst_methylseq">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">00_differentialMethylationSequencing</h1>
<h4 class="author">Caitlin Page</h4>
<h4 class="date">2024-10-25</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-04-11
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>2_ontology_bias/analysis/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240905code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240905)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240905code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240905)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomcaitlinpagegstmethylseqtreeae3cb7017b640c80a6ace84f6378765cb692553atargetblankae3cb70a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/caitlinpage/gst_methylseq/tree/ae3cb7017b640c80a6ace84f6378765cb692553a" target="_blank">ae3cb70</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomcaitlinpagegstmethylseqtreeae3cb7017b640c80a6ace84f6378765cb692553atargetblankae3cb70a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/caitlinpage/gst_methylseq/tree/ae3cb7017b640c80a6ace84f6378765cb692553a" target="_blank">ae3cb70</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/figure/
    Ignored:    analysis/input/
    Ignored:    analysis/mc_h1/
    Ignored:    analysis/myEnvironment.RData
    Ignored:    data/cell_line_wgbs/
    Ignored:    data/comp_bio_challenge/
    Ignored:    data/gene_sets/
    Ignored:    data/long_read/
    Ignored:    data/microarray/
    Ignored:    data/wgbs/
    Ignored:    output/figures/
    Ignored:    renv/library/
    Ignored:    renv/staging/

Untracked files:
    Untracked:  analysis/00_bugsLearning.Rmd
    Untracked:  analysis/04_swapInputs.Rmd
    Untracked:  analysis/05_vizResults.Rmd
    Untracked:  analysis/clusterlog.txt
    Untracked:  code/ChatGPT_translate.html
    Untracked:  code/ChatGPT_translate_files/
    Untracked:  code/background.R

Unstaged changes:
    Modified:   .gitignore
    Modified:   2_ontology_bias.Rproj
    Modified:   _workflowr.yml
    Modified:   analysis/00_exploreCpGGST.Rmd
    Modified:   analysis/03_genesToGST.Rmd
    Modified:   analysis/03_howToGST.Rmd
    Modified:   analysis/03_runGST.Rmd
    Modified:   analysis/04a_compareGSTResults.Rmd
    Modified:   analysis/06_decisionSummary.Rmd
    Modified:   analysis/index.Rmd
    Modified:   code/mmGST.R
    Modified:   code/runSeqDmMethods.R
    Modified:   renv.lock
    Modified:   renv/.gitignore
    Modified:   renv/activate.R
    Modified:   renv/settings.json

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/00_differentialMethylationSequencing.Rmd</code>) and
HTML (<code>docs/00_differentialMethylationSequencing.html</code>)
files. If you’ve configured a remote Git repository (see
<code>?wflow_git_remote</code>), click on the hyperlinks in the table
below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/ae3cb7017b640c80a6ace84f6378765cb692553a/analysis/00_differentialMethylationSequencing.Rmd" target="_blank">ae3cb70</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-04-11
</td>
<td>
wflow_publish("00_differentialMethylationSequencing.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/caitlinpage/gst_methylseq/67fb63b7b04f3976be5507b007139d70cf86649c/docs/00_differentialMethylationSequencing.html" target="_blank">67fb63b</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-04-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/50cb92f9b70f45f0c5f4365ad36ad31d441d1b63/analysis/00_differentialMethylationSequencing.Rmd" target="_blank">50cb92f</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-04-09
</td>
<td>
wflow_publish("00_differentialMethylationSequencing.Rmd")
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/f19ca4d6ea19a5321479feaadd63bfe13433513e/analysis/00_differentialMethylationSequencing.Rmd" target="_blank">f19ca4d</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-03-31
</td>
<td>
different methods
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/b149ec31f485f3e9881ab01a1572f66469654b11/analysis/00_differentialMethylationSequencing.Rmd" target="_blank">b149ec3</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-02-21
</td>
<td>
dmrcate dm - bc can also do for array
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/59852853edc6bebfb396fbb0a8e8a4d40d2e0aee/analysis/00_differentialMethylationSequencing.Rmd" target="_blank">5985285</a>
</td>
<td>
Caitlin Page
</td>
<td>
2024-11-12
</td>
<td>
init commit
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="data" class="section level2">
<h2>Data</h2>
<p>The data used in this analysis is B cell types and NK cell types from
WGBS, under the hg19 genome annotation. <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE186458">GSE186458</a></p>
<ul>
<li>B cell type: 3 samples</li>
<li>NK cell type: 3 samples</li>
</ul>
<p>Further description can be found <a
href="00_datasets.html">here</a></p>
</div>
<div id="packages-used" class="section level2">
<h2>Packages used</h2>
<pre class="r"><code>library(DMRcate)
library(bsseq)
library(edgeR)
library(limma)
library(BiocParallel)
library(BSgenome.Hsapiens.UCSC.hg19)
library(Biostrings)

library(plyranges)
library(dplyr)
library(ggplot2)
library(ggVennDiagram)
library(ComplexUpset)</code></pre>
</div>
<div id="description" class="section level2">
<h2>Description</h2>
<p>To understand the inputs of Gene Set Testing, I explored different
differential methylation methods. I focussed on methods available within
Bioconductor, and ones that were widely used. I tried the following
methods: * <a
href="https://www.bioconductor.org/packages/release/bioc/html/bsseq.html">bsseq</a>
<a
href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2012-13-10-r83">Paper</a>
Years in Bioc: 12.5</p>
<ul>
<li><p><a
href="https://www.bioconductor.org/packages/release/bioc/html/DMRcate.html">DMRcate</a>
<a
href="https://academic.oup.com/nar/article/49/19/e109/6329576?login=false">Paper</a>
Years in Bioc: 11</p></li>
<li><p><a
href="https://www.bioconductor.org/packages/release/bioc/html/dmrseq.html">dmrseq</a>
<a
href="https://academic.oup.com/biostatistics/article/20/3/367/4899074?login=false">Paper</a>
Years in Bioc: 7</p></li>
<li><p><a
href="https://www.bioconductor.org/packages/release/bioc/html/DSS.html">DSS</a>
<a
href="https://academic.oup.com/nar/article/42/8/e69/1074350?login=false">Paper</a>
Years in Bioc: 12.5</p></li>
<li><p><a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC5747346/">edgeR
workflow</a></p></li>
</ul>
</div>
<div id="obtain-counts" class="section level2">
<h2>Obtain counts</h2>
<pre class="r"><code># get the positions of cg sites
seq_names &lt;- seqnames(BSgenome.Hsapiens.UCSC.hg19)[1:25]
anno_seq &lt;- lapply(seq_names, function(x) {
  cbind(data.frame(matchPattern(&quot;CG&quot;, BSgenome.Hsapiens.UCSC.hg19[[x]])),
        seqnames = x)}) %&gt;%
  dplyr::bind_rows()
anno_seq &lt;- anno_seq %&gt;% mutate(pos = paste0(seqnames, &quot;-&quot;, start)) %&gt;%
  relocate(pos, seqnames)

# NK counts
fname &lt;- &quot;../data/wgbs/GSM5652299_Blood-NK-Z000000TM.beta&quot;
N &lt;- file.info(fname)$size
beta_nk &lt;- data.frame(matrix(readBin(fname, &quot;integer&quot;, N, size = 1, signed = FALSE), N / 2, 2, byrow=TRUE))
fname &lt;- &quot;../data/wgbs/GSM5652300_Blood-NK-Z000000U1.beta&quot;
N &lt;- file.info(fname)$size
beta_nk &lt;- cbind(beta_nk, data.frame(matrix(readBin(fname, &quot;integer&quot;, N, size = 1, signed = FALSE), N / 2, 2, byrow=TRUE)))
fname &lt;- &quot;../data/wgbs/GSM5652301_Blood-NK-Z000000UF.beta&quot;
N &lt;- file.info(fname)$size
beta_nk &lt;- cbind(beta_nk, data.frame(matrix(readBin(fname, &quot;integer&quot;, N, size = 1, signed = FALSE), N / 2, 2, byrow=TRUE)))
colnames(beta_nk) &lt;- c(&quot;M_TM&quot;, &quot;C_TM&quot;, &quot;M_U1&quot;, &quot;C_U1&quot;, &quot;M_UF&quot;, &quot;C_UF&quot;)
beta_nk &lt;- cbind(anno_seq[,1:4], beta_nk)

# B counts
fname &lt;- &quot;../data/wgbs/GSM5652316_Blood-B-Z000000TX.beta&quot;
N &lt;- file.info(fname)$size
beta_b &lt;- data.frame(matrix(readBin(fname, &quot;integer&quot;, N, size = 1, signed = FALSE), N / 2, 2, byrow=TRUE))
fname &lt;- &quot;../data/wgbs/GSM5652317_Blood-B-Z000000UB.beta&quot;
N &lt;- file.info(fname)$size
beta_b &lt;- cbind(beta_b, data.frame(matrix(readBin(fname, &quot;integer&quot;, N, size = 1, signed = FALSE), N / 2, 2, byrow=TRUE)))
fname &lt;- &quot;../data/wgbs/GSM5652318_Blood-B-Z000000UR.beta&quot;
N &lt;- file.info(fname)$size
beta_b &lt;- cbind(beta_b, data.frame(matrix(readBin(fname, &quot;integer&quot;, N, size = 1, signed = FALSE), N / 2, 2, byrow=TRUE)))
colnames(beta_b) &lt;- c(&quot;M_TX&quot;, &quot;C_TX&quot;, &quot;M_UB&quot;, &quot;C_UB&quot;, &quot;M_UR&quot;, &quot;C_UR&quot;)
beta_b &lt;- cbind(anno_seq[,1:4], beta_b)

# add unmeth counts
beta_nk &lt;- beta_nk %&gt;% mutate(Un_TM = C_TM - M_TM, Un_U1 = C_U1 - M_U1, Un_UF = C_UF - M_UF)
beta_b &lt;- beta_b %&gt;% mutate(Un_TX = C_TX - M_TX, Un_UB = C_UB - M_UB, Un_UR = C_UR - M_UR)

# combine
colnames(beta_b)[5:13] &lt;- paste0(&quot;b_&quot;, colnames(beta_b)[5:13])
colnames(beta_nk)[5:13] &lt;- paste0(&quot;nk_&quot;, colnames(beta_nk)[5:13])

wgbs_counts &lt;- cbind(beta_nk, beta_b[,5:13])

#saveRDS(wgbs_counts, &quot;../output/wgbs_counts.rds&quot;)</code></pre>
<pre class="r"><code>nrow(wgbs_counts)</code></pre>
<pre><code>[1] 28217448</code></pre>
<pre class="r"><code>head(wgbs_counts)</code></pre>
<pre><code>         pos seqnames start   end nk_M_TM nk_C_TM nk_M_U1 nk_C_U1 nk_M_UF
1 chr1-10469     chr1 10469 10470      27      32      41      48      21
2 chr1-10471     chr1 10471 10472      25      35      45      50      19
3 chr1-10484     chr1 10484 10485      29      35      42      50      26
4 chr1-10489     chr1 10489 10490      33      34      46      51      26
5 chr1-10493     chr1 10493 10494      29      36      49      53      28
6 chr1-10497     chr1 10497 10498      33      37      52      54      30
  nk_C_UF nk_Un_TM nk_Un_U1 nk_Un_UF b_M_TX b_C_TX b_M_UB b_C_UB b_M_UR b_C_UR
1      30        5        7        9     31     34     19     20     40     44
2      30       10        5       11     26     36     20     21     35     44
3      32        6        8        6     34     39     22     26     46     51
4      33        1        5        7     36     41     25     27     43     50
5      33        7        4        5     37     43     24     27     46     52
6      34        4        2        4     39     43     28     29     52     57
  b_Un_TX b_Un_UB b_Un_UR
1       3       1       4
2      10       1       9
3       5       4       5
4       5       2       7
5       6       3       6
6       4       1       5</code></pre>
</div>
<div id="differential-methylation-analysis" class="section level2">
<h2>Differential Methylation Analysis</h2>
</div>
<div id="bsseq-pre-processing-bsseq-and-bsmooth-objects"
class="section level1">
<h1>bsseq pre-processing: bsseq and bsmooth objects</h1>
<pre class="r"><code>## set up
wgbs_counts &lt;- wgbs_counts %&gt;% filter(!seqnames %in% c(&quot;chrX&quot;, &quot;chrY&quot;, &quot;chrM&quot;))
m_df &lt;- wgbs_counts[,c(5,7,9,14,16,18)]
colnames(m_df) &lt;- colnames(m_df) %&gt;% gsub(&quot;_M&quot;, &quot;&quot;, .)

coverage_df &lt;- wgbs_counts[,c(6,8,10,15,17,19)]
colnames(coverage_df) &lt;- colnames(coverage_df) %&gt;% gsub(&quot;_C&quot;, &quot;&quot;, .)

sample_stuff &lt;- cbind(Type = c(&quot;nk&quot;, &quot;nk&quot;, &quot;nk&quot;, &quot;b&quot;, &quot;b&quot;, &quot;b&quot;)) %&gt;% data.frame()
rownames(sample_stuff) &lt;- colnames(m_df)

## make BSseq object
bseq_nk_vs_b_all &lt;- BSseq(M = as.matrix(m_df),
                          Cov = as.matrix(coverage_df), pData = sample_stuff,
                          chr = beta_all$seqnames, pos = as.numeric(beta_all$start),
                          sampleNames = colnames(m_df))

## BSmooth by sample - so it doesn&#39;t crash
bsmooth1 &lt;- BSmooth(BSseq = bseq_nk_vs_b_all[,1], BPPARAM = MulticoreParam(workers = 6), verbose = TRUE)

bsmooth2 &lt;- BSmooth(BSseq = bseq_nk_vs_b_all[,2], BPPARAM = MulticoreParam(workers = 6), verbose = TRUE)
bsmooth3 &lt;- BSmooth(BSseq = bseq_nk_vs_b_all[,3], BPPARAM = MulticoreParam(workers = 6), verbose = TRUE)

bsmooth4 &lt;- BSmooth(BSseq = bseq_nk_vs_b_all[,4], BPPARAM = MulticoreParam(workers = 6), verbose = TRUE)
bsmooth5 &lt;- BSmooth(BSseq = bseq_nk_vs_b_all[,5], BPPARAM = MulticoreParam(workers = 6), verbose = TRUE)
bsmooth6 &lt;- BSmooth(BSseq = bseq_nk_vs_b_all[,6], BPPARAM = MulticoreParam(workers = 6), verbose = TRUE)

bsmooth_obj &lt;- combine(bsmooth1, bsmooth2, bsmooth3, bsmooth4, bsmooth5, bsmooth6)
## filter
bseq.cov &lt;- getCoverage(bsmooth_obj)
keep &lt;- which(rowSums(bseq.cov[, bsmooth_obj$Type == &quot;nk&quot;] &gt;= 2) &gt;= 2 &amp;
                rowSums(bseq.cov[, bsmooth_obj$Type == &quot;b&quot;] &gt;= 2) &gt;= 2)
length(keep)

bsmooth_obj &lt;- bsmooth_obj[keep,]
saveRDS(bsmooth_obj, &quot;../output/bsmooth_obj.rds&quot;)</code></pre>
</div>
<div id="bsseq" class="section level1">
<h1>bsseq</h1>
<pre class="r"><code>## testing
bsseq.tstat &lt;- BSmooth.tstat(bsmooth_obj, group1 = c(&quot;nk_TM&quot;, &quot;nk_U1&quot;, &quot;nk_UF&quot;),
                            group2 = c(&quot;b_TX&quot;, &quot;b_UB&quot;, &quot;b_UR&quot;),
                            estimate.var = &quot;group2&quot;,
                            local.correct = TRUE,
                            verbose = TRUE, mc.cores = 6)

bsseq_res &lt;- bsseq.tstat@stats %&gt;% data.frame() %&gt;%
  mutate(seqnames = data.frame(bsseq.tstat@gr)$seqnames,
         start = data.frame(bsseq.tstat@gr)$start,
         end = start, position = paste0(seqnames, &quot;-&quot;, start)) %&gt;%
  relocate(position)
saveRDS(bsseq_res, &quot;../output/bsseq_res.rds&quot;)

## dmrs
bsseq_dmrs &lt;- dmrFinder(bseq.tstat)
# filter the dmrs
bsseq_dmrs &lt;- subset(bsseq_dmrs, n &gt;= 3 &amp; abs(meanDiff) &gt;= 0.1)
colnames(bsseq_dmrs)[1] &lt;- &quot;seqnames&quot;
bsseq_dmrs &lt;- bsseq_dmrs %&gt;%
  mutate(position = paste0(seqnames, &quot;-&quot;, start)) %&gt;% relocate(position)
bsseq_dmrs$rank &lt;- 1:nrow(bsseq_dmrs)
saveRDS(bsseq_dmrs, &quot;../output/bsseq_dmrs.rds&quot;)</code></pre>
<pre class="r"><code>nrow(bsseq_res)</code></pre>
<pre><code>[1] 26141562</code></pre>
<pre class="r"><code>head(bsseq_res)</code></pre>
<pre><code>    position     rawSds   tstat.sd group2.means group1.means     tstat
1 chr1-10469 0.03810513 0.02830937    0.8974952    0.8654874 -1.130643
2 chr1-10471 0.03767339 0.02932521    0.8979726    0.8648112 -1.130814
3 chr1-10484 0.03509357 0.02993758    0.9008921    0.8603429 -1.354458
4 chr1-10489 0.03420441 0.03027659    0.9019333    0.8586008 -1.431223
5 chr1-10493 0.03353366 0.03054339    0.9027350    0.8572023 -1.490753
6 chr1-10497 0.03289861 0.03078513    0.9035093    0.8558021 -1.549685
  tstat.corrected seqnames start   end
1      -0.2440643     chr1 10469 10469
2      -0.2445427     chr1 10471 10471
3      -0.4701852     chr1 10484 10484
4      -0.5477191     chr1 10489 10489
5      -0.6078640     chr1 10493 10493
6      -0.6674112     chr1 10497 10497</code></pre>
<pre class="r"><code>nrow(bsseq_dmrs)</code></pre>
<pre><code>[1] 69858</code></pre>
<pre class="r"><code>head(bsseq_dmrs)</code></pre>
<pre><code>            position seqnames     start       end idxStart   idxEnd cluster   n
90156 chr7-158595973     chr7 158595973 158597702 11824761 11824875 2008166 115
97889   chr11-314783    chr11    314783    318111 15588281 15588431  287130 151
68822   chr2-8418248     chr2   8418248   8424108  2314827  2314938  918827 112
56322  chr19-1646759    chr19   1646759   1650002 23525162 23525269  889718 108
54340 chr17-75316243    chr17  75316243  75321648 22648912 22649108  823284 197
51278 chr16-88600393    chr16  88600393  88601461 21620676 21620832  779278 157
      width invdensity  areaStat   maxStat   meanDiff group1.mean group2.mean
90156  1730  15.043478 -2448.388 -2.608966 -0.7669851   0.1469680   0.9139531
97889  3329  22.046358 -2381.623 -3.101464 -0.6354995   0.0890310   0.7245305
68822  5861  52.330357 -2374.264 -4.032202 -0.6338283   0.2448161   0.8786444
56322  3244  30.037037  1998.382 24.989896  0.6892998   0.9072248   0.2179250
54340  5406  27.441624  1987.636 25.101076  0.3506880   0.7423341   0.3916461
51278  1069   6.808917  1985.715 16.149314  0.3906955   0.5600847   0.1693891
        tstat.sd direction
90156 0.03370352      hypo
97889 0.02951290      hypo
68822 0.03179347      hypo
56322 0.03597023     hyper
54340 0.03016071     hyper
51278 0.02959988     hyper</code></pre>
</div>
<div id="dmrcate" class="section level1">
<h1>DMRcate</h1>
<ul>
<li>takes in smoothed samples from bsseq</li>
<li>also has the same filtering recommendations as bsseq</li>
</ul>
<pre class="r"><code>bsmooth_obj

tissue &lt;- factor(c(&quot;nk&quot;, &quot;nk&quot;, &quot;nk&quot;, &quot;b&quot;, &quot;b&quot;, &quot;b&quot;), levels = c(&quot;nk&quot;, &quot;b&quot;))
tissue
#Regular matrix design
design &lt;- model.matrix(~tissue)
design

methdesign &lt;- modelMatrixMeth(design)
methdesign

seq_annot &lt;- sequencing.annotate(bsmooth_obj, methdesign, all.cov = TRUE, 
                                 coef = 8, fdr=0.05)
#saveRDS(seq_annot, &quot;../output/dmrcate_seq_anno.rds&quot;)

seq_dmrcate.res &lt;- dmrcate(seq_annot, C=2, min.cpgs = 5)

dmrcate_seq_dmr &lt;- extractRanges(seq_dmrcate.res, genome = &quot;hg19&quot;)
dmrcate_seq_dmr &lt;- dmrcate_seq_dmr %&gt;% data.frame()
dmrcate_seq_anno &lt;- seq_annot@ranges %&gt;% data.frame()
#saveRDS(dmrcate_seq_dmr, &quot;../output/dmrcate_seq_dmr.rds&quot;)</code></pre>
<pre class="r"><code>nrow(dmrcate_seq_anno)</code></pre>
<pre><code>[1] 25995712</code></pre>
<pre class="r"><code>head(dmrcate_seq_anno)</code></pre>
<pre><code>  seqnames start   end width strand       stat       diff   ind.fdr is.sig
1     chr1 10469 10469     1      * -1.8140361 -1.3572918 0.6992461  FALSE
2     chr1 10471 10471     1      * -0.7284539 -0.6399003 0.9389270  FALSE
3     chr1 10484 10484     1      * -0.7000213 -0.5063787 0.9425177  FALSE
4     chr1 10489 10489     1      *  0.3100487  0.2549802 0.9833920  FALSE
5     chr1 10493 10493     1      * -0.1626419 -0.1219368 0.9929045  FALSE
6     chr1 10497 10497     1      * -0.2503823 -0.1957677 0.9876618  FALSE</code></pre>
<pre class="r"><code>nrow(dmrcate_seq_dmr)</code></pre>
<pre><code>[1] 13542</code></pre>
<pre class="r"><code>head(dmrcate_seq_dmr)</code></pre>
<pre><code>  seqnames     start       end width strand no.cpgs min_smoothed_fdr
1     chr1 167434091 167488800 54710      *     715                0
2    chr11 113932115 113948338 16224      *     336                0
3    chr19   8627507   8641759 14253      *     316                0
4    chr12  10456885  10466669  9785      *     127                0
5    chr10  72355543  72364759  9217      *     198                0
6    chr17  45811354  45822029 10676      *     208                0
       Stouffer        HMFDR        Fisher    maxdiff  meandiff
1  0.000000e+00 0.0002481331  0.000000e+00 -11.311593 -5.643222
2  0.000000e+00 0.0007797373 7.262418e-315  -9.825363 -4.674082
3 7.545644e-243 0.0005360353 1.003690e-278 -10.136485 -4.386791
4 3.058706e-290 0.0001100756 4.325037e-267 -10.604822 -7.088274
5 1.123502e-263 0.0002704979 4.449882e-261  -9.472313 -5.302931
6 3.710160e-216 0.0005478491 2.540144e-232  -9.936832 -4.962632
  overlapping.genes
1   AKR1D1P1, CD247
2            ZBTB16
3             MYO1F
4             KLRD1
5              PRF1
6             TBX21</code></pre>
</div>
<div id="dmrseq" class="section level1">
<h1>dmrseq</h1>
<ul>
<li>Uses bsseq object as input</li>
<li>dmrseq failed.</li>
</ul>
<pre class="r"><code>bseq_nk_vs_b_all

loci.idx &lt;- which(DelayedMatrixStats::rowSums2(getCoverage(bseq_nk_vs_b_all, type=&quot;Cov&quot;)==0) == 0)
sample.idx &lt;- which(pData(bseq_nk_vs_b_all)$Type %in% c(&quot;nk&quot;, &quot;b&quot;))

bs.filtered &lt;- bseq_nk_vs_b_all[loci.idx, sample.idx]

testCovariate &lt;- &quot;Type&quot;
regions &lt;- dmrseq(bs=bs.filtered,
                  cutoff = 0.05,
                  testCovariate=testCovariate) # failed with error</code></pre>
</div>
<div id="dss" class="section level1">
<h1>DSS</h1>
<ul>
<li>Uses bsseq object as input</li>
<li>DMLtest failed to finish. Unclear as to why, expected to be too
computationally intensive</li>
</ul>
<pre class="r"><code>bseq_nk_vs_b_all
dss_dmlTest_all &lt;- DMLtest(bseq_nk_vs_b_all,
                                  group1=c(&quot;nk_TM&quot;, &quot;nk_U1&quot;, &quot;nk_UF&quot;),
                                  group2 = c(&quot;b_TX&quot;, &quot;b_UB&quot;, &quot;b_UR&quot;),
                               smoothing = TRUE, BPPARAM = BatchtoolsParam()) #failed
dss_dmrs_pos_filt &lt;- callDMR(dss_dmlTest_nk_vs_b_pos_filt, p.threshold=0.05)
dss_dmrs &lt;- callDMR(dss_dmlTest_nk_vs_b, p.threshold=0.05)</code></pre>
</div>
<div id="edger" class="section level1">
<h1>edgeR</h1>
<pre class="r"><code># pre-processing
edge_matrix &lt;- wgbs_counts[,c(5,11,7,12,9,13,14,20,16,21,18,22)]

colnames(edge_matrix) &lt;- c(
  &quot;NK1.Me&quot;, &quot;NK1.Un&quot;, &quot;NK2.Me&quot;, &quot;NK2.Un&quot;, &quot;NK3.Me&quot;, &quot;NK3.Un&quot;,
  &quot;B1.Me&quot;, &quot;B1.Un&quot;, &quot;B2.Me&quot;, &quot;B2.Un&quot;, &quot;B3.Me&quot;, &quot;B3.Un&quot;)
rownames(edge_matrix) &lt;- wgbs_counts$pos

edge_matrix &lt;- edge_matrix[!grepl(&quot;chrX&quot;, rownames(edge_matrix)),]
edge_matrix &lt;- edge_matrix[!grepl(&quot;chrY&quot;, rownames(edge_matrix)),]

sample &lt;- gl(6,2,12)
methylation &lt;- gl(2,1,12, labels=c(&quot;Me&quot;,&quot;Un&quot;))
condition &lt;- gl(2,6,12, labels=c(&quot;NK&quot;, &quot;B&quot;))
edge_sample_info &lt;- cbind(sample, methylation, condition)
rownames(edge_sample_info) &lt;- colnames(edge_matrix)
data.frame(edge_sample_info)


edge_dge &lt;- DGEList(as.matrix(edge_matrix), samples = data.frame(edge_sample_info),
                    group = c(
                      &quot;NK&quot;, &quot;NK&quot;, &quot;NK&quot;, &quot;NK&quot;, &quot;NK&quot;, &quot;NK&quot;,
                      &quot;B&quot;, &quot;B&quot;, &quot;B&quot;, &quot;B&quot;, &quot;B&quot;, &quot;B&quot;))

edge_coverage &lt;- edge_dge$counts[, c(1,3,5,7,9,11)] + edge_dge$counts[, c(2,4,6,8,10,12)]
head(edge_coverage)
keep &lt;- rowSums(edge_coverage &gt;= 8) == 6

edge_dge &lt;- edge_dge[keep,,keep.lib.sizes=FALSE]

edge_dge$samples

edge_total_lib_size &lt;- filter(edge_dge$samples, methylation==1)$lib.size +
  filter(edge_dge$samples, methylation==2)$lib.size

edge_dge$samples$lib.size &lt;- rep(edge_total_lib_size, each=2)
edge_dge$samples

# testing
edge_samples &lt;- cbind(Group = c(&quot;NK&quot;, &quot;NK&quot;, &quot;NK&quot;, &quot;B&quot;, &quot;B&quot;, &quot;B&quot;)) %&gt;% data.frame()
rownames(edge_samples) &lt;- c(&quot;NK1&quot;, &quot;NK2&quot;, &quot;NK3&quot;, &quot;B1&quot;, &quot;B2&quot;, &quot;B3&quot;)
edge_samples &lt;- edge_samples %&gt;% mutate(Group = as.factor(Group))

edge_design &lt;- model.matrix(~0 + Group, data = edge_samples)
colnames(edge_design) &lt;- c(&quot;B&quot;, &quot;NK&quot;)

edge_design &lt;- modelMatrixMeth(edge_design)

edge_dge &lt;- estimateDisp(edge_dge, edge_design, trend=&quot;none&quot;)

edge_fit &lt;- glmFit(edge_dge, edge_design)
edge_contrasts &lt;- makeContrasts(NK - B,
                                levels=edge_design)
edge_lrt &lt;- glmLRT(edge_fit, contrast=edge_contrasts)

edge_res &lt;- topTags(edge_lrt, n = Inf)

edge_res &lt;- edge_res$table %&gt;% mutate(signif.05 = ifelse(FDR &lt;= 0.05, TRUE, FALSE))

#saveRDS(edge_res, &quot;../output/edge_res.rds&quot;)</code></pre>
<pre class="r"><code>nrow(edge_res)</code></pre>
<pre><code>[1] 24287284</code></pre>
<pre class="r"><code>head(edge_res)</code></pre>
<pre><code>                   logFC    logCPM       LR       PValue          FDR signif.05
chr2-102985195 -14.57712 -4.792479 129.4411 5.430423e-30 1.318902e-22      TRUE
chr8-60100608  -15.06021 -4.923078 127.5833 1.384675e-29 1.681499e-22      TRUE
chr22-18077698  14.74929 -5.060096 120.3964 5.180342e-28 2.369875e-21      TRUE
chr5-156608059 -14.72724 -5.081965 120.2161 5.673161e-28 2.369875e-21      TRUE
chr12-10432700 -14.70074 -5.075146 120.2159 5.673586e-28 2.369875e-21      TRUE
chr7-127645445  14.72284 -5.078940 120.1536 5.854606e-28 2.369875e-21      TRUE</code></pre>
<div id="compare-approaches" class="section level2">
<h2>Compare approaches</h2>
<ul>
<li>All methods were run on the HPC at PeterMac because they are very
computationally intense.</li>
<li>Despite this, DSS and dmrseq still failed - unclear as to why, so we
moved on.</li>
<li>The edgeR workflow does not identify DMRs</li>
</ul>
<pre class="r"><code>compare_dm &lt;- wgbs_counts[,1:4]
bsseq_dmrs_filt &lt;- bsseq_dmrs %&gt;% filter(abs(areaStat) &gt;= 100)
overlap_dm &lt;- find_overlaps(as_granges(dmrcate_seq_dmr[1:5000,]), as_granges(compare_dm)) %&gt;% data.frame()
compare_dm &lt;- compare_dm %&gt;% mutate(seq_dmrcate_dmr_5k = ifelse(pos %in% overlap_dm$pos, TRUE, FALSE))
overlap_dm &lt;- find_overlaps(as_granges(bsseq_dmrs_filt), as_granges(compare_dm)) %&gt;% data.frame()
compare_dm &lt;- compare_dm %&gt;% mutate(bsseq_dmr_area100 = ifelse(pos %in% overlap_dm$pos, TRUE, FALSE))

compare_dm &lt;- compare_dm %&gt;% mutate(edger_pos = edge_res[match(.$pos, rownames(edge_res)), &quot;signif.05&quot;])</code></pre>
<pre class="r"><code>upset(filter(compare_dm, seq_dmrcate_dmr_5k != FALSE| bsseq_dmr_area100 != FALSE| edger_pos != FALSE), intersect = colnames(compare_dm)[5:7])</code></pre>
<pre><code>Warning in upset_data(data, intersect, mode = mode, encode_sets = encode_sets,
: Detected missing values in the columns indicating sets, coercing to FALSE</code></pre>
<p><img src="figure/00_differentialMethylationSequencing.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="compare-bsseq-and-dmrcate" class="section level1">
<h1>Compare bsseq and DMRCate</h1>
<pre class="r"><code>rbind(cbind(width = bsseq_dmrs$width, method = &quot;bsseq&quot;), 
      cbind(width = dmrcate_seq_dmr$width, method = &quot;dmrcate&quot;)) %&gt;% data.frame() %&gt;% 
  mutate(width = as.double(width)) %&gt;%
  ggplot(aes(x = log2(width), fill = method)) +
  geom_density(alpha = 0.4) +
  theme_bw()</code></pre>
<p><img src="figure/00_differentialMethylationSequencing.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>rbind(cbind(width = dmrcate_seq_dmr$width, method = &quot;dmrcate&quot;),
      cbind(width = dmrcate_seq_dmr[1:5000,]$width, method = &quot;dmrcate_5k&quot;),
      cbind(width = dmrcate_seq_dmr[1:1000,]$width, method = &quot;dmrcate_1k&quot;),
      cbind(width = bsseq_dmrs_filt$width, method = &quot;bsseq_area100&quot;)) %&gt;% data.frame() %&gt;% mutate(width = as.double(width)) %&gt;%
  ggplot(aes(x = width, colour = method)) +
  geom_density()</code></pre>
<p><img src="figure/00_differentialMethylationSequencing.Rmd/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>rbind(cbind(width = dmrcate_seq_dmr$width, method = &quot;dmrcate&quot;),
      cbind(width = dmrcate_seq_dmr[1:5000,]$width, method = &quot;dmrcate_5k&quot;),
      cbind(width = dmrcate_seq_dmr[1:1000,]$width, method = &quot;dmrcate_1k&quot;),
      cbind(width = bsseq_dmrs_filt$width, method = &quot;bsseq_area100&quot;)) %&gt;% data.frame() %&gt;% mutate(width = as.double(width)) %&gt;%
  ggplot(aes(x = log2(width), colour = method)) +
  geom_density()</code></pre>
<p><img src="figure/00_differentialMethylationSequencing.Rmd/unnamed-chunk-22-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>rbind(cbind(no.cpgs = dmrcate_seq_dmr$no.cpgs, method = &quot;dmrcate&quot;),
      cbind(no.cpgs = dmrcate_seq_dmr[1:5000,]$no.cpgs, method = &quot;dmrcate_5k&quot;),
      cbind(no.cpgs = dmrcate_seq_dmr[1:1000,]$no.cpgs, method = &quot;dmrcate_1k&quot;),
      cbind(no.cpgs = bsseq_dmrs_filt$n, method = &quot;bsseq_area100&quot;)) %&gt;% data.frame() %&gt;% mutate(no.cpgs = as.double(no.cpgs)) %&gt;%
  ggplot(aes(x = no.cpgs, colour = method)) +
  geom_density()</code></pre>
<p><img src="figure/00_differentialMethylationSequencing.Rmd/unnamed-chunk-23-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>rbind(cbind(no.cpgs = dmrcate_seq_dmr$no.cpgs, method = &quot;dmrcate&quot;),
      cbind(no.cpgs = dmrcate_seq_dmr[1:5000,]$no.cpgs, method = &quot;dmrcate_5k&quot;),
      cbind(no.cpgs = dmrcate_seq_dmr[1:1000,]$no.cpgs, method = &quot;dmrcate_1k&quot;),
      cbind(no.cpgs = bsseq_dmrs_filt$n, method = &quot;bsseq_area100&quot;)) %&gt;% data.frame() %&gt;% mutate(no.cpgs = as.double(no.cpgs)) %&gt;%
  ggplot(aes(x = log2(no.cpgs), colour = method)) +
  geom_density()</code></pre>
<p><img src="figure/00_differentialMethylationSequencing.Rmd/unnamed-chunk-23-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>cg_in_dmr &lt;- wgbs_counts[,c(&quot;pos&quot;, &quot;seqnames&quot;, &quot;start&quot;)]
cg_in_dmr$end &lt;- cg_in_dmr$start
overlap &lt;- find_overlaps(as_granges(cg_in_dmr), as_granges(bsseq_dmrs_filt)) %&gt;% data.frame()
cg_in_dmr$bsseq &lt;- ifelse(cg_in_dmr$pos %in% overlap$pos, TRUE, FALSE)

overlap &lt;- find_overlaps(as_granges(cg_in_dmr), as_granges(dmrcate_seq_dmr)) %&gt;% data.frame()
cg_in_dmr$dmrcate &lt;- ifelse(cg_in_dmr$pos %in% overlap$pos, TRUE, FALSE)</code></pre>
<pre class="r"><code>ggVennDiagram(list(bsseq = filter(cg_in_dmr, bsseq == TRUE)$pos, dmrcate = filter(cg_in_dmr, dmrcate == TRUE)$pos))</code></pre>
<p><img src="figure/00_differentialMethylationSequencing.Rmd/unnamed-chunk-25-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggVennDiagram(list(bsseq = filter(cg_in_dmr, bsseq != TRUE)$pos, dmrcate = filter(cg_in_dmr, dmrcate != TRUE)$pos))</code></pre>
<p><img src="figure/00_differentialMethylationSequencing.Rmd/unnamed-chunk-25-2.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.2

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Melbourne
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] ComplexUpset_1.3.3                ggVennDiagram_1.5.2              
 [3] ggplot2_3.5.1                     dplyr_1.1.4                      
 [5] plyranges_1.24.0                  BSgenome.Hsapiens.UCSC.hg19_1.4.3
 [7] BSgenome_1.72.0                   rtracklayer_1.64.0               
 [9] BiocIO_1.14.0                     Biostrings_2.72.1                
[11] XVector_0.44.0                    BiocParallel_1.38.0              
[13] edgeR_4.2.2                       limma_3.60.6                     
[15] bsseq_1.40.0                      SummarizedExperiment_1.34.0      
[17] Biobase_2.64.0                    MatrixGenerics_1.16.0            
[19] matrixStats_1.5.0                 GenomicRanges_1.56.2             
[21] GenomeInfoDb_1.40.1               IRanges_2.38.1                   
[23] S4Vectors_0.42.1                  BiocGenerics_0.50.0              
[25] DMRcate_3.0.10                   

loaded via a namespace (and not attached):
  [1] fs_1.6.5                                           
  [2] ProtGenerics_1.36.0                                
  [3] bitops_1.0-9                                       
  [4] httr_1.4.7                                         
  [5] RColorBrewer_1.1-3                                 
  [6] tools_4.4.2                                        
  [7] doRNG_1.8.6.1                                      
  [8] backports_1.5.0                                    
  [9] R6_2.5.1                                           
 [10] HDF5Array_1.32.1                                   
 [11] lazyeval_0.2.2                                     
 [12] Gviz_1.48.0                                        
 [13] rhdf5filters_1.16.0                                
 [14] permute_0.9-7                                      
 [15] withr_3.0.2                                        
 [16] prettyunits_1.2.0                                  
 [17] gridExtra_2.3                                      
 [18] base64_2.0.2                                       
 [19] preprocessCore_1.66.0                              
 [20] cli_3.6.3                                          
 [21] labeling_0.4.3                                     
 [22] sass_0.4.9                                         
 [23] IlluminaHumanMethylationEPICanno.ilm10b4.hg19_0.6.0
 [24] readr_2.1.5                                        
 [25] genefilter_1.86.0                                  
 [26] askpass_1.2.1                                      
 [27] Rsamtools_2.20.0                                   
 [28] foreign_0.8-87                                     
 [29] siggenes_1.78.0                                    
 [30] illuminaio_0.46.0                                  
 [31] R.utils_2.12.3                                     
 [32] dichromat_2.0-0.1                                  
 [33] scrime_1.3.5                                       
 [34] rstudioapi_0.17.1                                  
 [35] RSQLite_2.3.9                                      
 [36] generics_0.1.3                                     
 [37] gtools_3.9.5                                       
 [38] Matrix_1.7-1                                       
 [39] interp_1.1-6                                       
 [40] abind_1.4-8                                        
 [41] R.methodsS3_1.8.2                                  
 [42] lifecycle_1.0.4                                    
 [43] whisker_0.4.1                                      
 [44] yaml_2.3.10                                        
 [45] rhdf5_2.48.0                                       
 [46] SparseArray_1.4.8                                  
 [47] BiocFileCache_2.12.0                               
 [48] grid_4.4.2                                         
 [49] blob_1.2.4                                         
 [50] promises_1.3.2                                     
 [51] ExperimentHub_2.12.0                               
 [52] crayon_1.5.3                                       
 [53] lattice_0.22-6                                     
 [54] GenomicFeatures_1.56.0                             
 [55] annotate_1.82.0                                    
 [56] KEGGREST_1.44.1                                    
 [57] pillar_1.10.1                                      
 [58] knitr_1.49                                         
 [59] beanplot_1.3.1                                     
 [60] rjson_0.2.23                                       
 [61] codetools_0.2-20                                   
 [62] glue_1.8.0                                         
 [63] data.table_1.16.4                                  
 [64] vctrs_0.6.5                                        
 [65] png_0.1-8                                          
 [66] gtable_0.3.6                                       
 [67] IlluminaHumanMethylation450kanno.ilmn12.hg19_0.6.1 
 [68] cachem_1.1.0                                       
 [69] xfun_0.50                                          
 [70] S4Arrays_1.4.1                                     
 [71] survival_3.7-0                                     
 [72] iterators_1.0.14                                   
 [73] statmod_1.5.0                                      
 [74] nlme_3.1-166                                       
 [75] bit64_4.6.0-1                                      
 [76] progress_1.2.3                                     
 [77] filelock_1.0.3                                     
 [78] rprojroot_2.0.4                                    
 [79] bslib_0.9.0                                        
 [80] nor1mix_1.3-3                                      
 [81] rpart_4.1.23                                       
 [82] colorspace_2.1-1                                   
 [83] DBI_1.2.3                                          
 [84] Hmisc_5.2-2                                        
 [85] nnet_7.3-19                                        
 [86] tidyselect_1.2.1                                   
 [87] bit_4.5.0.1                                        
 [88] compiler_4.4.2                                     
 [89] curl_6.2.0                                         
 [90] git2r_0.35.0                                       
 [91] httr2_1.1.0                                        
 [92] htmlTable_2.4.3                                    
 [93] xml2_1.3.6                                         
 [94] DelayedArray_0.30.1                                
 [95] checkmate_2.3.2                                    
 [96] scales_1.3.0                                       
 [97] quadprog_1.5-8                                     
 [98] rappdirs_0.3.3                                     
 [99] stringr_1.5.1                                      
[100] digest_0.6.37                                      
[101] rmarkdown_2.29                                     
[102] GEOquery_2.72.0                                    
[103] htmltools_0.5.8.1                                  
[104] pkgconfig_2.0.3                                    
[105] jpeg_0.1-10                                        
[106] base64enc_0.1-3                                    
[107] sparseMatrixStats_1.16.0                           
[108] dbplyr_2.5.0                                       
[109] fastmap_1.2.0                                      
[110] ensembldb_2.28.1                                   
[111] rlang_1.1.5                                        
[112] htmlwidgets_1.6.4                                  
[113] UCSC.utils_1.0.0                                   
[114] DelayedMatrixStats_1.26.0                          
[115] farver_2.1.2                                       
[116] jquerylib_0.1.4                                    
[117] jsonlite_1.8.9                                     
[118] mclust_6.1.1                                       
[119] R.oo_1.27.0                                        
[120] VariantAnnotation_1.50.0                           
[121] RCurl_1.98-1.16                                    
[122] magrittr_2.0.3                                     
[123] Formula_1.2-5                                      
[124] GenomeInfoDbData_1.2.12                            
[125] patchwork_1.3.0                                    
[126] Rhdf5lib_1.26.0                                    
[127] munsell_0.5.1                                      
[128] Rcpp_1.0.14                                        
[129] stringi_1.8.4                                      
[130] zlibbioc_1.50.0                                    
[131] MASS_7.3-61                                        
[132] AnnotationHub_3.12.0                               
[133] plyr_1.8.9                                         
[134] bumphunter_1.46.0                                  
[135] org.Hs.eg.db_3.19.1                                
[136] minfi_1.50.0                                       
[137] parallel_4.4.2                                     
[138] deldir_2.0-4                                       
[139] splines_4.4.2                                      
[140] multtest_2.60.0                                    
[141] hms_1.1.3                                          
[142] locfit_1.5-9.11                                    
[143] rngtools_1.5.2                                     
[144] biomaRt_2.60.1                                     
[145] BiocVersion_3.19.1                                 
[146] missMethyl_1.38.0                                  
[147] XML_3.99-0.18                                      
[148] evaluate_1.0.3                                     
[149] latticeExtra_0.6-30                                
[150] biovizBase_1.52.0                                  
[151] BiocManager_1.30.25                                
[152] tzdb_0.4.0                                         
[153] foreach_1.5.2                                      
[154] httpuv_1.6.15                                      
[155] tidyr_1.3.1                                        
[156] openssl_2.3.2                                      
[157] purrr_1.0.4                                        
[158] reshape_0.8.9                                      
[159] xtable_1.8-4                                       
[160] restfulr_0.0.15                                    
[161] AnnotationFilter_1.28.0                            
[162] later_1.4.1                                        
[163] tibble_3.2.1                                       
[164] memoise_2.0.1                                      
[165] AnnotationDbi_1.66.0                               
[166] GenomicAlignments_1.40.0                           
[167] cluster_2.1.6                                      
[168] workflowr_1.7.1                                    </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
