<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Caitlin Page" />

<meta name="date" content="2025-02-27" />

<title>03_runGST</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">2_ontology_bias</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/caitlinpage/gst_methylseq">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">03_runGST</h1>
<h4 class="author">Caitlin Page</h4>
<h4 class="date">2025-02-27</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-05-07
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>2_ontology_bias/analysis/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240905code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240905)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240905code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240905)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomcaitlinpagegstmethylseqtree90ba1164575823b550ac57d5d7c6f1cf63f2109ctargetblank90ba116a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/caitlinpage/gst_methylseq/tree/90ba1164575823b550ac57d5d7c6f1cf63f2109c" target="_blank">90ba116</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomcaitlinpagegstmethylseqtree90ba1164575823b550ac57d5d7c6f1cf63f2109ctargetblank90ba116a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/caitlinpage/gst_methylseq/tree/90ba1164575823b550ac57d5d7c6f1cf63f2109c" target="_blank">90ba116</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/input/
    Ignored:    analysis/mc_h1/
    Ignored:    analysis/myEnvironment.RData
    Ignored:    data/cell_line_wgbs/
    Ignored:    data/gene_sets/
    Ignored:    data/long_read/
    Ignored:    data/microarray/
    Ignored:    data/wgbs/
    Ignored:    data/wgbs_2/
    Ignored:    data/wgbs_3/
    Ignored:    output/figures/
    Ignored:    renv/library/
    Ignored:    renv/staging/

Untracked files:
    Untracked:  analysis/00_cpgFeatures.Rmd
    Untracked:  analysis/04_compareGSTResults.Rmd
    Untracked:  analysis/04_promoterAnalysis.Rmd
    Untracked:  analysis/07_monoGranAnalysis.Rmd
    Untracked:  analysis/07_tCellAnalysis.Rmd
    Untracked:  analysis/clusterlog.txt
    Untracked:  analysis/input.tar.gz
    Untracked:  analysis/myEnvironment.RDataTmp
    Untracked:  code/ChatGPT_translate.html
    Untracked:  code/ChatGPT_translate_files/
    Untracked:  code/background.R
    Untracked:  code/dm_analysis_t.R

Unstaged changes:
    Modified:   .gitignore
    Modified:   2_ontology_bias.Rproj
    Modified:   _workflowr.yml
    Modified:   analysis/00_datasets.Rmd
    Modified:   analysis/00_longReadAnalysis.Rmd
    Modified:   analysis/03_genesToGST.Rmd
    Modified:   analysis/04_swapInputs.Rmd
    Modified:   analysis/05_plots.Rmd
    Modified:   analysis/index.Rmd
    Modified:   code/function_binBiasPlot.R
    Modified:   renv.lock
    Modified:   renv/.gitignore
    Modified:   renv/activate.R
    Modified:   renv/settings.json

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/03_runGST.Rmd</code>) and HTML
(<code>docs/03_runGST.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/90ba1164575823b550ac57d5d7c6f1cf63f2109c/analysis/03_runGST.Rmd" target="_blank">90ba116</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-05-07
</td>
<td>
wflow_publish("03_runGST.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/caitlinpage/gst_methylseq/2997135671348905d11e2a33ae99a725d041a059/docs/03_runGST.html" target="_blank">2997135</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-05-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/d8755b4c76ae6b884f6ec6ef3e70101b95dcae79/analysis/03_runGST.Rmd" target="_blank">d8755b4</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-05-01
</td>
<td>
wflow_publish("03_runGST.Rmd")
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/604fd47443c30fb4765c765cc9ee939236fd1153/analysis/03_runGST.Rmd" target="_blank">604fd47</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-04-23
</td>
<td>
update
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/d5121f67f1fa85534e88da0d31ff819667e24263/analysis/03_runGST.Rmd" target="_blank">d5121f6</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-03-31
</td>
<td>
results
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/4ff80cdc638011806ea453acc956adef7ed9e54f/analysis/03_runGST.Rmd" target="_blank">4ff80cd</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-03-13
</td>
<td>
run gst with dmrcate
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<pre class="r"><code>library(plyranges)</code></pre>
<pre><code>Loading required package: BiocGenerics</code></pre>
<pre><code>
Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    IQR, mad, sd, var, xtabs</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, table,
    tapply, union, unique, unsplit, which.max, which.min</code></pre>
<pre><code>Loading required package: IRanges</code></pre>
<pre><code>Loading required package: S4Vectors</code></pre>
<pre><code>Loading required package: stats4</code></pre>
<pre><code>
Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:utils&#39;:

    findMatches</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    expand.grid, I, unname</code></pre>
<pre><code>Loading required package: GenomicRanges</code></pre>
<pre><code>Loading required package: GenomeInfoDb</code></pre>
<pre><code>
Attaching package: &#39;plyranges&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:IRanges&#39;:

    slice</code></pre>
<pre><code>The following object is masked from &#39;package:stats&#39;:

    filter</code></pre>
<pre class="r"><code>library(dplyr)</code></pre>
<pre><code>
Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:plyranges&#39;:

    between, n, n_distinct</code></pre>
<pre><code>The following objects are masked from &#39;package:GenomicRanges&#39;:

    intersect, setdiff, union</code></pre>
<pre><code>The following object is masked from &#39;package:GenomeInfoDb&#39;:

    intersect</code></pre>
<pre><code>The following objects are masked from &#39;package:IRanges&#39;:

    collapse, desc, intersect, setdiff, slice, union</code></pre>
<pre><code>The following objects are masked from &#39;package:S4Vectors&#39;:

    first, intersect, rename, setdiff, setequal, union</code></pre>
<pre><code>The following objects are masked from &#39;package:BiocGenerics&#39;:

    combine, intersect, setdiff, union</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    filter, lag</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    intersect, setdiff, setequal, union</code></pre>
<pre class="r"><code>library(ggplot2)

library(BiasedUrn)
library(ggVennDiagram)
library(org.Hs.eg.db)</code></pre>
<pre><code>Loading required package: AnnotationDbi</code></pre>
<pre><code>Loading required package: Biobase</code></pre>
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    &#39;browseVignettes()&#39;. To cite Bioconductor, see
    &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>
Attaching package: &#39;AnnotationDbi&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:dplyr&#39;:

    select</code></pre>
<pre><code>The following object is masked from &#39;package:plyranges&#39;:

    select</code></pre>
<pre><code></code></pre>
<pre class="r"><code>wgbs_counts &lt;- readRDS(&quot;../output/wgbs_counts.rds&quot;)
biomart_genes &lt;- readRDS(&quot;../output/genes_biomart.rds&quot;)
dmrcate_seq_dmr &lt;- readRDS(&quot;../output/dmrcate_seq_dmr.rds&quot;) %&gt;% data.frame()
dmrcate_seq_anno &lt;- readRDS(&quot;../output/dmrcate_seq_anno.rds&quot;)
dmrcate_seq_anno &lt;- data.frame(dmrcate_seq_anno@ranges)

immuneGO &lt;- read.table(&quot;../data/gene_sets/GO-immune-system-process.txt&quot;, header = FALSE)
names(immuneGO) &lt;- &quot;GOID&quot;
immuneGO &lt;- distinct(immuneGO)</code></pre>
<div id="selected-bias" class="section level1">
<h1>selected bias</h1>
<ul>
<li>Number of CpGs overlapping a Gene</li>
</ul>
<div id="run-sequencing" class="section level2">
<h2>run sequencing</h2>
<pre class="r"><code>source(&quot;../code/mmGST.R&quot;)
source(&quot;../code/function_binBiasPlot.R&quot;)</code></pre>
<pre><code>Loading required package: GenomicAlignments</code></pre>
<pre><code>Loading required package: SummarizedExperiment</code></pre>
<pre><code>Loading required package: MatrixGenerics</code></pre>
<pre><code>Loading required package: matrixStats</code></pre>
<pre><code>
Attaching package: &#39;matrixStats&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:Biobase&#39;:

    anyMissing, rowMedians</code></pre>
<pre><code>The following object is masked from &#39;package:dplyr&#39;:

    count</code></pre>
<pre><code>
Attaching package: &#39;MatrixGenerics&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:matrixStats&#39;:

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
<pre><code>The following object is masked from &#39;package:Biobase&#39;:

    rowMedians</code></pre>
<pre><code>Loading required package: Biostrings</code></pre>
<pre><code>Loading required package: XVector</code></pre>
<pre><code>
Attaching package: &#39;Biostrings&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:base&#39;:

    strsplit</code></pre>
<pre><code>Loading required package: Rsamtools</code></pre>
<pre><code>
Attaching package: &#39;GenomicAlignments&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:dplyr&#39;:

    last</code></pre>
<pre class="r"><code>gst_res_dmr1 &lt;- run_miss_methyl_b(run_miss_methyl_a(biomart_genes, annoGeneDmr(dmrcate_seq_anno, dmrcate_seq_dmr[1:1000,], &quot;biomart&quot;, biomart_genes)), &quot;Wallenius&quot;)[[1]]</code></pre>
<pre><code>&#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="figure/03_runGST.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-3-1">
Past versions of unnamed-chunk-3-1.png
</button>
</p>
<div id="fig-unnamed-chunk-3-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/2997135671348905d11e2a33ae99a725d041a059/docs/figure/03_runGST.Rmd/unnamed-chunk-3-1.png" target="_blank">2997135</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>gst_res_dmr1$rank_fdr &lt;- 1:nrow(gst_res_dmr1)
gst_res_dmr1 &lt;- gst_res_dmr1 %&gt;% mutate(is.sig = ifelse(FDR &lt; 0.05, TRUE, FALSE))
head(gst_res_dmr1)</code></pre>
<pre><code>                 GOID ONTOLOGY                  TERM     N  DE         P.DE
GO:0002376 GO:0002376       BP immune system process  2489 180 2.806219e-43
GO:0006955 GO:0006955       BP       immune response  1778 142 3.067599e-40
GO:0045321 GO:0045321       BP  leukocyte activation   887  94 1.008091e-31
GO:0005488 GO:0005488       MF               binding 14490 499 3.739115e-30
GO:0001775 GO:0001775       BP       cell activation  1027  97 1.140039e-28
GO:0046649 GO:0046649       BP lymphocyte activation   736  81 2.001881e-28
                    FDR SigGenesInSet rank_fdr is.sig
GO:0002376 2.078566e-39            NA        1   TRUE
GO:0006955 1.136085e-36            NA        2   TRUE
GO:0045321 2.488976e-28            NA        3   TRUE
GO:0005488 6.923907e-27            NA        4   TRUE
GO:0001775 1.688854e-25            NA        5   TRUE
GO:0046649 2.471323e-25            NA        6   TRUE</code></pre>
<pre class="r"><code>#saveRDS(gst_res_dmr1, &quot;../output/gst_res_dmrcate_1k.rds&quot;)</code></pre>
<pre class="r"><code>gst_res_dmr1[1:10,] %&gt;% mutate(in_imm = ifelse(GOID %in% immuneGO$GOID, TRUE, FALSE)) %&gt;% .[order(.$FDR, decreasing = TRUE),] %&gt;%
  ggplot(aes(x = -log10(FDR), y = factor(.data$TERM, levels = .data$TERM), colour = in_imm, size = N)) +
  geom_point() +
  geom_vline(xintercept = 0.05, type = &quot;dotted&quot;) </code></pre>
<pre><code>Warning in geom_vline(xintercept = 0.05, type = &quot;dotted&quot;): Ignoring unknown
parameters: `type`</code></pre>
<p><img src="figure/03_runGST.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-4-1">
Past versions of unnamed-chunk-4-1.png
</button>
</p>
<div id="fig-unnamed-chunk-4-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/2997135671348905d11e2a33ae99a725d041a059/docs/figure/03_runGST.Rmd/unnamed-chunk-4-1.png" target="_blank">2997135</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="impact-of-dmr-filtering-parameters" class="section level2">
<h2>Impact of DMR filtering parameters</h2>
<pre class="r"><code>source(&quot;../code/mmGST.R&quot;)
source(&quot;../code/function_binBiasPlot.R&quot;)
gst_res_all &lt;- run_miss_methyl_b(run_miss_methyl_a(biomart_genes, annoGeneDmr(wgbs_counts, dmrcate_seq_dmr, &quot;biomart&quot;, biomart_genes)), &quot;Wallenius&quot;)[[1]]
gst_res_all$rank_fdr &lt;- 1:nrow(gst_res_all)
gst_res_all &lt;- gst_res_all %&gt;% mutate(is.sig = ifelse(FDR &lt; 0.05, TRUE, FALSE))

gst_res_dmr5 &lt;- run_miss_methyl_b(run_miss_methyl_a(biomart_genes, annoGeneDmr(wgbs_counts, dmrcate_seq_dmr[1:5000,], &quot;biomart&quot;, biomart_genes)), &quot;Wallenius&quot;)[[1]]
gst_res_dmr5$rank_fdr &lt;- 1:nrow(gst_res_dmr5)
gst_res_dmr5 &lt;- gst_res_dmr5 %&gt;% mutate(is.sig = ifelse(FDR &lt; 0.05, TRUE, FALSE))

#saveRDS(gst_res_dmr5, &quot;../output/gst_res_dmrcate_5k.rds&quot;)
#saveRDS(gst_res_all, &quot;../output/gst_res_dmrcate_all.rds&quot;)</code></pre>
<pre class="r"><code>head(gst_res_dmr5)</code></pre>
<pre><code>                 GOID ONTOLOGY                       TERM     N   DE
GO:0003674 GO:0003674       MF         molecular_function 15958 1871
GO:0005488 GO:0005488       MF                    binding 14490 1769
GO:0110165 GO:0110165       CC cellular anatomical entity 16581 1887
GO:0008150 GO:0008150       BP         biological_process 16081 1844
GO:0005575 GO:0005575       CC         cellular_component 16841 1894
GO:0009987 GO:0009987       BP           cellular process 15088 1771
                   P.DE          FDR SigGenesInSet rank_fdr is.sig
GO:0003674 7.899195e-77 9.670195e-73            NA        1   TRUE
GO:0005488 1.572736e-75 9.626720e-72            NA        2   TRUE
GO:0110165 5.398275e-74 2.202856e-70            NA        3   TRUE
GO:0008150 4.103695e-73 1.255936e-69            NA        4   TRUE
GO:0005575 3.423553e-72 8.382228e-69            NA        5   TRUE
GO:0009987 1.393786e-71 2.843788e-68            NA        6   TRUE</code></pre>
<pre class="r"><code>head(gst_res_all)</code></pre>
<pre><code>                 GOID ONTOLOGY                       TERM     N   DE
GO:0008150 GO:0008150       BP         biological_process 16081 3663
GO:0003674 GO:0003674       MF         molecular_function 15958 3698
GO:0005575 GO:0005575       CC         cellular_component 16841 3764
GO:0110165 GO:0110165       CC cellular anatomical entity 16581 3740
GO:0009987 GO:0009987       BP           cellular process 15088 3498
GO:0005488 GO:0005488       MF                    binding 14490 3468
                    P.DE           FDR SigGenesInSet rank_fdr is.sig
GO:0008150 1.199295e-152 1.847034e-148            NA        1   TRUE
GO:0003674 1.069463e-150 8.235402e-147            NA        2   TRUE
GO:0005575 3.055545e-149 1.568615e-145            NA        3   TRUE
GO:0110165 2.786500e-148 1.072872e-144            NA        4   TRUE
GO:0009987 9.976731e-143 3.073033e-139            NA        5   TRUE
GO:0005488 1.104056e-140 2.833929e-137            NA        6   TRUE</code></pre>
<pre class="r"><code>venn_list &lt;- list(
  dmr_all = gst_res_all %&gt;% filter(FDR &lt; 0.05) %&gt;% .$GOID,
  dmr_5k = gst_res_dmr5 %&gt;% filter(FDR &lt; 0.05) %&gt;% .$GOID,
  dmr_1k = gst_res_dmr1 %&gt;% filter(FDR &lt; 0.05) %&gt;% .$GOID
)
ggVennDiagram(venn_list)</code></pre>
<p><img src="figure/03_runGST.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-8-1">
Past versions of unnamed-chunk-8-1.png
</button>
</p>
<div id="fig-unnamed-chunk-8-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/2997135671348905d11e2a33ae99a725d041a059/docs/figure/03_runGST.Rmd/unnamed-chunk-8-1.png" target="_blank">2997135</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="wallenius-vs-fisher" class="section level2">
<h2>Wallenius vs Fisher</h2>
<pre class="r"><code>source(&quot;../code/mmGST.R&quot;)
source(&quot;../code/function_binBiasPlot.R&quot;)
gst_res_dmr1_fish &lt;- run_miss_methyl_b(run_miss_methyl_a(biomart_genes, annoGeneDmr(wgbs_counts, dmrcate_seq_dmr[1:1000,], &quot;biomart&quot;, biomart_genes)), &quot;Fishers&quot;)[[1]]
gst_res_dmr1_fish$rank_fdr &lt;- 1:nrow(gst_res_dmr1_fish)
gst_res_dmr1_fish &lt;- gst_res_dmr1_fish %&gt;% mutate(is.sig = ifelse(FDR &lt; 0.05, TRUE, FALSE))
saveRDS(gst_res_dmr1_fish, &quot;../output/gst_res_dmr1_fish.rds&quot;)</code></pre>
<pre class="r"><code>head(gst_res_dmr1_fish)</code></pre>
<pre><code>                 GOID ONTOLOGY                  TERM     N  DE         P.DE
GO:0002376 GO:0002376       BP immune system process  2489 180 2.180085e-43
GO:0006955 GO:0006955       BP       immune response  1778 142 2.653965e-40
GO:0045321 GO:0045321       BP  leukocyte activation   887  94 8.179942e-32
GO:0005488 GO:0005488       MF               binding 14490 499 1.792719e-30
GO:0001775 GO:0001775       BP       cell activation  1027  97 9.234773e-29
GO:0046649 GO:0046649       BP lymphocyte activation   736  81 1.669776e-28
                    FDR SigGenesInSet rank_fdr is.sig
GO:0002376 1.614789e-39            NA        1   TRUE
GO:0006955 9.828959e-37            NA        2   TRUE
GO:0045321 2.019628e-28            NA        3   TRUE
GO:0005488 3.319668e-27            NA        4   TRUE
GO:0001775 1.368039e-25            NA        5   TRUE
GO:0046649 2.061338e-25            NA        6   TRUE</code></pre>
<pre class="r"><code>source(&quot;../code/mmGST.R&quot;)
source(&quot;../code/function_binBiasPlot.R&quot;)
system.time(run_miss_methyl_b(run_miss_methyl_a(biomart_genes, annoGeneDmr(wgbs_counts, dmrcate_seq_dmr[1:100,], &quot;biomart&quot;, biomart_genes)), &quot;Fishers&quot;))</code></pre>
<pre><code>&#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="figure/03_runGST.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-12-1">
Past versions of unnamed-chunk-12-1.png
</button>
</p>
<div id="fig-unnamed-chunk-12-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/2997135671348905d11e2a33ae99a725d041a059/docs/figure/03_runGST.Rmd/unnamed-chunk-12-1.png" target="_blank">2997135</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre><code>   user  system elapsed 
117.998   5.044 123.946 </code></pre>
<pre class="r"><code>system.time(run_miss_methyl_b(run_miss_methyl_a(biomart_genes, annoGeneDmr(wgbs_counts, dmrcate_seq_dmr[1:100,], &quot;biomart&quot;, biomart_genes)), &quot;Wallenius&quot;))</code></pre>
<pre><code>&#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<pre><code>   user  system elapsed 
117.612   4.498 122.973 </code></pre>
<pre class="r"><code>venn_list &lt;- list(wall = filter(gst_res_dmr1, FDR &lt; 0.05)$TERM, fish = filter(gst_res_dmr1_fish, FDR &lt; 0.05)$TERM)
ggVennDiagram(venn_list)</code></pre>
<p><img src="figure/03_runGST.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-13-1">
Past versions of unnamed-chunk-13-1.png
</button>
</p>
<div id="fig-unnamed-chunk-13-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/2997135671348905d11e2a33ae99a725d041a059/docs/figure/03_runGST.Rmd/unnamed-chunk-13-1.png" target="_blank">2997135</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="wallenius-vs-hypergeometric-unbiased-vs-biased"
class="section level2">
<h2>Wallenius vs Hypergeometric (Unbiased vs Biased)</h2>
<pre class="r"><code>source(&quot;../code/mmGST.R&quot;)
source(&quot;../code/function_binBiasPlot.R&quot;)
gst_res_dmr1_hyper &lt;- run_miss_methyl_b(run_miss_methyl_a(biomart_genes, annoGeneDmr(wgbs_counts, dmrcate_seq_dmr[1:1000,], &quot;biomart&quot;, biomart_genes)), &quot;None&quot;)[[1]]
gst_res_dmr1_hyper$rank_fdr &lt;- 1:nrow(gst_res_dmr1_hyper)
gst_res_dmr1_hyper &lt;- gst_res_dmr1_hyper %&gt;% mutate(is.sig = ifelse(FDR &lt; 0.05, TRUE, FALSE))

saveRDS(gst_res_dmr1_hyper, &quot;../output/gst_res_dmr1_hyper.rds&quot;)</code></pre>
<pre class="r"><code>head(gst_res_dmr1_hyper)</code></pre>
<pre><code>                 GOID ONTOLOGY                       TERM     N  DE
GO:0005488 GO:0005488       MF                    binding 14490 499
GO:0003674 GO:0003674       MF         molecular_function 15958 518
GO:0110165 GO:0110165       CC cellular anatomical entity 16581 523
GO:0005575 GO:0005575       CC         cellular_component 16841 524
GO:0008150 GO:0008150       BP         biological_process 16081 508
GO:0005515 GO:0005515       MF            protein binding 12312 445
                   P.DE          FDR SigGenesInSet rank_fdr is.sig
GO:0005488 3.131759e-74 2.319694e-70            NA        1   TRUE
GO:0003674 3.907231e-72 1.447043e-68            NA        2   TRUE
GO:0110165 5.230662e-69 1.291451e-65            NA        3   TRUE
GO:0005575 9.119677e-67 1.688736e-63            NA        4   TRUE
GO:0008150 5.555703e-62 8.230218e-59            NA        5   TRUE
GO:0005515 1.057193e-61 1.305104e-58            NA        6   TRUE</code></pre>
<pre class="r"><code>venn_list &lt;- list(wall = filter(gst_res_dmr1, FDR &lt; 0.05)$TERM, hyper = filter(gst_res_dmr1_hyper, FDR &lt; 0.05)$TERM)
ggVennDiagram(venn_list)</code></pre>
<p><img src="figure/03_runGST.Rmd/unnamed-chunk-17-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-17-1">
Past versions of unnamed-chunk-17-1.png
</button>
</p>
<div id="fig-unnamed-chunk-17-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/2997135671348905d11e2a33ae99a725d041a059/docs/figure/03_runGST.Rmd/unnamed-chunk-17-1.png" target="_blank">2997135</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="comparisons" class="section level2">
<h2>Comparisons</h2>
<pre class="r"><code>xx &lt;- as.list(org.Hs.egGO2ALLEGS)
out &lt;- run_miss_methyl_a(biomart_genes, annoGeneDmr(wgbs_counts, dmrcate_seq_dmr[1:500,], &quot;biomart&quot;, biomart_genes))</code></pre>
<pre><code>&#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<pre class="r"><code>go_entrez &lt;- data.frame(cbind(go = gsub(&quot;\\..*&quot;, &quot;&quot;, names(unlist(xx))), entrez = unlist(xx)))
go_entrez &lt;- go_entrez[,1:2] %&gt;% data.frame()
go_entrez &lt;- go_entrez %&gt;%
  mutate(num_cg = data.frame(out$freq)[match(.$entrez, data.frame(out$freq)$freq_make_table), &quot;Freq&quot;],
         dmrcate_1k = ifelse(go %in% rownames(filter(gst_res_dmr1, FDR &lt; 0.05)), TRUE, FALSE),
         dmrcate_1k_null = ifelse(go %in% rownames(filter(gst_res_dmr1_hyper, FDR &lt; 0.05)), TRUE, FALSE),
         dmrcate_1k_fish = ifelse(go %in% rownames(filter(gst_res_dmr1_fish, FDR &lt; 0.05)), TRUE, FALSE),
         dmrcate_5k = ifelse(go %in% rownames(filter(gst_res_dmr5, FDR &lt; 0.05)), TRUE, FALSE),
         dmrcate_all = ifelse(go %in% rownames(filter(gst_res_all, FDR &lt; 0.05)), TRUE, FALSE))
go_entrez &lt;- go_entrez %&gt;% mutate(num_cg = ifelse(is.na(num_cg), 0, num_cg))
go_entrez &lt;- go_entrez %&gt;% group_by(go) %&gt;% mutate(mean_cg = mean(num_cg)) %&gt;% ungroup()

go_entrez &lt;- go_entrez %&gt;%
  mutate(rank_1k = gst_res_dmr1[match(.$go, gst_res_dmr1$GOID), &quot;rank_fdr&quot;],
         rank_1k_null = gst_res_dmr1_hyper[match(.$go, gst_res_dmr1_hyper$GOID), &quot;rank_fdr&quot;],
         rank_1k_fish = gst_res_dmr1_fish[match(.$go, gst_res_dmr1_fish$GOID), &quot;rank_fdr&quot;],
         rank_5k = gst_res_dmr5[match(.$go, gst_res_dmr5$GOID), &quot;rank_fdr&quot;],
         rank_all = gst_res_all[match(.$go, gst_res_all$GOID), &quot;rank_fdr&quot;])</code></pre>
<pre class="r"><code>go_entrez %&gt;% distinct(go, mean_cg, dmrcate_1k, dmrcate_1k_null, rank_1k, rank_1k_null) %&gt;%
  filter(rank_1k &lt;= 1000, rank_1k_null &lt;= 1000) %&gt;%
  ggplot(aes(x = mean_cg, y = rank_1k_null - rank_1k)) + 
  geom_point() +
  geom_hline(yintercept = 0, colour = &quot;red&quot;, linetype = &quot;dashed&quot;) +
  labs(x = &quot;Average No. CpGs in GO category&quot;, y = &quot;Increase in rank in Wallenius&quot;) +
  scale_x_continuous(limits = c(0, 1500))</code></pre>
<pre><code>Warning: Removed 20 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
<p><img src="figure/03_runGST.Rmd/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-19-1">
Past versions of unnamed-chunk-19-1.png
</button>
</p>
<div id="fig-unnamed-chunk-19-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/2997135671348905d11e2a33ae99a725d041a059/docs/figure/03_runGST.Rmd/unnamed-chunk-19-1.png" target="_blank">2997135</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>go_entrez %&gt;% distinct(go, mean_cg, dmrcate_1k, dmrcate_1k_null) %&gt;% 
  filter(dmrcate_1k_null == TRUE | dmrcate_1k == TRUE) %&gt;% 
  mutate(Method = case_when(dmrcate_1k == FALSE &amp; dmrcate_1k_null == TRUE ~ &quot;Biased&quot;,
                            dmrcate_1k == TRUE &amp; dmrcate_1k_null == TRUE ~ &quot;Both&quot;,
                            TRUE ~ &quot;Bias Corrected&quot;)) %&gt;%
  ggplot(aes(x = reorder(Method, mean_cg), y = mean_cg, fill = Method)) +
  geom_boxplot() +
  theme_bw() +
  labs(x = &quot;Method&quot;, y = &quot;Average number of CpGs per ontology category&quot;)</code></pre>
<p><img src="figure/03_runGST.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-20-1">
Past versions of unnamed-chunk-20-1.png
</button>
</p>
<div id="fig-unnamed-chunk-20-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/2997135671348905d11e2a33ae99a725d041a059/docs/figure/03_runGST.Rmd/unnamed-chunk-20-1.png" target="_blank">2997135</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>truth_gst_100 &lt;- rbind(
  gst_res_dmr1[1:100,] %&gt;% mutate(in_imm = cumsum(GOID %in% immuneGO$GOID), method = &quot;Wall_1&quot;) %&gt;% .[names(.) %in% c(&quot;rank_fdr&quot;, &quot;in_imm&quot;, &quot;method&quot;)],
  gst_res_dmr1_hyper[1:100,] %&gt;% mutate(in_imm = cumsum(GOID %in% immuneGO$GOID), method = &quot;Hyper_1&quot;) %&gt;% .[names(.) %in% c(&quot;rank_fdr&quot;, &quot;in_imm&quot;, &quot;method&quot;)])

truth_gst_100 %&gt;%
  ggplot(aes(x = rank_fdr, y = in_imm, colour = method)) +
  geom_line(size = 2) +
  labs(x = &quot;Rank&quot;, y = &quot;Cumulative no. in immune sets&quot;) +
  theme_bw()</code></pre>
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
generated.</code></pre>
<p><img src="figure/03_runGST.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-21-1">
Past versions of unnamed-chunk-21-1.png
</button>
</p>
<div id="fig-unnamed-chunk-21-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/2997135671348905d11e2a33ae99a725d041a059/docs/figure/03_runGST.Rmd/unnamed-chunk-21-1.png" target="_blank">2997135</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="section" class="section level4">
<h4></h4>
<pre class="r"><code>c(cpgs_500dmrs = dmrcate_seq_dmr[1:500,] %&gt;% summarise(sum(no.cpgs)),
  cpgs_1kdmrs = dmrcate_seq_dmr[1:1000,] %&gt;% summarise(sum(no.cpgs)),
  cpgs_5kdmrs = dmrcate_seq_dmr[1:5000,] %&gt;% summarise(sum(no.cpgs)),
  cpgs_alldmrs = dmrcate_seq_dmr %&gt;% summarise(sum(no.cpgs)))</code></pre>
<pre><code>$`cpgs_500dmrs.sum(no.cpgs)`
[1] 44817

$`cpgs_1kdmrs.sum(no.cpgs)`
[1] 71184

$`cpgs_5kdmrs.sum(no.cpgs)`
[1] 188976

$`cpgs_alldmrs.sum(no.cpgs)`
[1] 344610</code></pre>
<pre class="r"><code>annoGeneDmr(wgbs_counts, dmrcate_seq_dmr[1:500,], &quot;biomart&quot;, biomart_genes) %&gt;% distinct(gene_name, num_cg) %&gt;% mutate(sum(num_cg))
annoGeneDmr(wgbs_counts, dmrcate_seq_dmr[1:1000,], &quot;biomart&quot;, biomart_genes) %&gt;% distinct(gene_name, num_cg) %&gt;% mutate(sum(num_cg))
annoGeneDmr(wgbs_counts, dmrcate_seq_dmr[1:5000,], &quot;biomart&quot;, biomart_genes) %&gt;% distinct(gene_name, num_cg) %&gt;% mutate(sum(num_cg))
annoGeneDmr(wgbs_counts, dmrcate_seq_dmr, &quot;biomart&quot;, biomart_genes) %&gt;% distinct(gene_name, num_cg) %&gt;% mutate(sum(num_cg))</code></pre>
<pre class="r"><code>dmrcate_seq_dmr %&gt;%
  ggplot(aes(x = log2(min_smoothed_fdr))) +
  geom_density()</code></pre>
<pre><code>Warning: Removed 494 rows containing non-finite outside the scale range
(`stat_density()`).</code></pre>
<p><img src="figure/03_runGST.Rmd/unnamed-chunk-24-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-24-1">
Past versions of unnamed-chunk-24-1.png
</button>
</p>
<div id="fig-unnamed-chunk-24-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/caitlinpage/gst_methylseq/blob/2997135671348905d11e2a33ae99a725d041a059/docs/figure/03_runGST.Rmd/unnamed-chunk-24-1.png" target="_blank">2997135</a>
</td>
<td>
Caitlin Page
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.2

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Melbourne
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] biomaRt_2.60.1              ChIPseqSpikeInFree_1.2.4   
 [3] GenomicAlignments_1.40.0    Rsamtools_2.20.0           
 [5] Biostrings_2.72.1           XVector_0.44.0             
 [7] SummarizedExperiment_1.34.0 MatrixGenerics_1.16.0      
 [9] matrixStats_1.5.0           reshape2_1.4.4             
[11] org.Hs.eg.db_3.19.1         AnnotationDbi_1.66.0       
[13] Biobase_2.64.0              ggVennDiagram_1.5.2        
[15] BiasedUrn_2.0.12            ggplot2_3.5.1              
[17] dplyr_1.1.4                 plyranges_1.24.0           
[19] GenomicRanges_1.56.2        GenomeInfoDb_1.40.1        
[21] IRanges_2.38.1              S4Vectors_0.42.1           
[23] BiocGenerics_0.50.0        

loaded via a namespace (and not attached):
 [1] DBI_1.2.3               bitops_1.0-9            httr2_1.1.0            
 [4] rlang_1.1.5             magrittr_2.0.3          git2r_0.35.0           
 [7] compiler_4.4.2          RSQLite_2.3.9           png_0.1-8              
[10] vctrs_0.6.5             stringr_1.5.1           pkgconfig_2.0.3        
[13] crayon_1.5.3            fastmap_1.2.0           dbplyr_2.5.0           
[16] labeling_0.4.3          promises_1.3.2          rmarkdown_2.29         
[19] UCSC.utils_1.0.0        bit_4.5.0.1             xfun_0.50              
[22] zlibbioc_1.50.0         cachem_1.1.0            jsonlite_1.8.9         
[25] progress_1.2.3          blob_1.2.4              later_1.4.1            
[28] DelayedArray_0.30.1     BiocParallel_1.38.0     parallel_4.4.2         
[31] prettyunits_1.2.0       R6_2.5.1                bslib_0.9.0            
[34] stringi_1.8.4           limma_3.60.6            rtracklayer_1.64.0     
[37] jquerylib_0.1.4         Rcpp_1.0.14             knitr_1.49             
[40] httpuv_1.6.15           Matrix_1.7-1            tidyselect_1.2.1       
[43] rstudioapi_0.17.1       abind_1.4-8             yaml_2.3.10            
[46] codetools_0.2-20        curl_6.2.0              lattice_0.22-6         
[49] tibble_3.2.1            plyr_1.8.9              withr_3.0.2            
[52] KEGGREST_1.44.1         evaluate_1.0.3          BiocFileCache_2.12.0   
[55] xml2_1.3.6              filelock_1.0.3          pillar_1.10.1          
[58] whisker_0.4.1           generics_0.1.3          rprojroot_2.0.4        
[61] RCurl_1.98-1.16         hms_1.1.3               munsell_0.5.1          
[64] scales_1.3.0            glue_1.8.0              tools_4.4.2            
[67] BiocIO_1.14.0           fs_1.6.5                XML_3.99-0.18          
[70] grid_4.4.2              colorspace_2.1-1        GenomeInfoDbData_1.2.12
[73] restfulr_0.0.15         cli_3.6.3               rappdirs_0.3.3         
[76] workflowr_1.7.1         S4Arrays_1.4.1          gtable_0.3.6           
[79] sass_0.4.9              digest_0.6.37           SparseArray_1.4.8      
[82] farver_2.1.2            rjson_0.2.23            memoise_2.0.1          
[85] htmltools_0.5.8.1       lifecycle_1.0.4         httr_1.4.7             
[88] statmod_1.5.0           GO.db_3.19.1            bit64_4.6.0-1          </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
