---
title: "00_differentialMethylationSequencing"
author: "Caitlin Page"
date: "2024-10-25"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: inline
---

## Introduction

```{r}

```

# do DMRcate tp match array
```{r}
library(DMRcate)

library(bsseq)
library(BiocParallel)
```

* they take in smoothed samples from bsseq
```{r}
wgbs_counts
```

```{r}
beta_all <- wgbs_counts
beta_all <- beta_all %>% filter(!seqnames %in% c("chrX", "chrY", "chrM"))
m_df <- beta_all[,c(5,7,9,14,16,18)]
colnames(m_df) <- colnames(m_df) %>% gsub("_M", "", .)

coverage_df <- beta_all[,c(6,8,10,15,17,19)]
colnames(coverage_df) <- colnames(coverage_df) %>% gsub("_C", "", .)

sample_stuff <- cbind(Type = c("nk", "nk", "nk", "b", "b", "b")) %>% data.frame()
rownames(sample_stuff) <- colnames(m_df)

## make BSseq object
bseq_nk_vs_b_all <- BSseq(M = as.matrix(m_df),
                          Cov = as.matrix(coverage_df), pData = sample_stuff,
                          chr = beta_all$seqnames, pos = as.numeric(beta_all$start),
                          sampleNames = colnames(m_df))
```

```{r}
#bsmooth_obj <- BSmooth(BSseq = bseq_nk_vs_b_all, BPPARAM = MulticoreParam(workers = 12), verbose = TRUE)
bsmooth_obj <- readRDS("../../1_methods/output/imp/bsmooth_all_obj.rds")
```
```{r}
bseq.cov <- getCoverage(bsmooth_obj)
keep <- which(rowSums(bseq.cov[, bsmooth_obj$Type == "nk"] >= 2) >= 2 &
                rowSums(bseq.cov[, bsmooth_obj$Type == "b"] >= 2) >= 2)
length(keep)

bsmooth_obj <- bsmooth_obj[keep,]
```

# now dmrcate
```{r}
colnames(bsmooth_obj)
```

```{r}
bsseq::pData(bsmooth_obj)
```


```{r}
tissue <- factor(pData(bsmooth_obj)$Type)
tissue
#Regular matrix design
design <- model.matrix(~tissue)
design
```

```{r}
methdesign <- edgeR::modelMatrixMeth(design)
methdesign
```
```{r}
saveRDS(bsmooth_obj, "../output/bsmooth_obj.rds")
```

```{r}
seq_annot <- sequencing.annotate(bsmooth_obj, methdesign, all.cov = TRUE,
coef = 8, fdr=0.05)
```
- well damn can only get around that on cluster
- but do i remember my cluster password? - probs not
- let the laptop calm down and then try again
- no that didn't work

```{r}
?sequencing.annotate
```

```{r}
seq_dmrcate.res <- dmrcate(seq_annot, C=2, min.cpgs = 5)
```

```{r}
seq_ranges <- extractRanges(seq_dmrcate.res, genome = "hg19")
```

############################


```{r}
library(DMRcate)
library(dplyr)
```
```{r}
library(bsseq)
library(BiocParallel)
library(limma)
```

```{r}
wgbs_counts <- readRDS("/researchers/caitlin.page/phd_gst/output/wgbs_counts.rds")
```


```{r}
beta_all <- wgbs_counts
beta_all <- beta_all %>% filter(!seqnames %in% c("chrX", "chrY", "chrM"))
m_df <- beta_all[,c(5,7,9,14,16,18)]
colnames(m_df) <- colnames(m_df) %>% gsub("_M", "", .)

coverage_df <- beta_all[,c(6,8,10,15,17,19)]
colnames(coverage_df) <- colnames(coverage_df) %>% gsub("_C", "", .)

sample_stuff <- cbind(Type = c("nk", "nk", "nk", "b", "b", "b")) %>% data.frame()
rownames(sample_stuff) <- colnames(m_df)

## make BSseq object
bseq_nk_vs_b_all <- BSseq(M = as.matrix(m_df),
                          Cov = as.matrix(coverage_df), pData = sample_stuff,
                          chr = beta_all$seqnames, pos = as.numeric(beta_all$start),
                          sampleNames = colnames(m_df))
```

```{r}
bsmooth_obj <- BSmooth(BSseq = bseq_nk_vs_b_all, BPPARAM = MulticoreParam(workers = 12), verbose = TRUE)
```
```{r}
bseq.cov <- getCoverage(bsmooth_obj)
keep <- which(rowSums(bseq.cov[, bsmooth_obj$Type == "nk"] >= 2) >= 2 &
                rowSums(bseq.cov[, bsmooth_obj$Type == "b"] >= 2) >= 2)
length(keep)

bsmooth_obj <- bsmooth_obj[keep,]
```

# dmrcate

```{r}
tissue <- factor(pData(bsmooth_obj)$Type)
tissue
#Regular matrix design
design <- model.matrix(~tissue)
design
```

```{r}
methdesign <- edgeR::modelMatrixMeth(design)
methdesign
```

```{r}
seq_annot <- sequencing.annotate(bsmooth_obj, methdesign, all.cov = TRUE,
coef = 8, fdr=0.05)
```

```{r}
dmrcate.res <- dmrcate(seq_annot, C=2, min.cpgs = 5)
```

```{r}
dmrcate_seq_dmr <-  extractRanges(dmrcate.res, genome="hg19")
```

