---
title: "00_exploreCpGDistrib"
author: "Caitlin Page"
date: "2024-09-05"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: inline
---

## Packages
```{r}
library(plyranges)
library(dplyr)
library(ggplot2)
library(ggVennDiagram)

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
library(IlluminaHumanMethylationEPICmanifest)
library(annotatr)
```

```{r}
biomart_genes %>% distinct(gene_name, num_cg) %>% group_by(num_cg) %>% summarise(frequency = n())
biomart_genes %>% distinct(gene_name, num_cg) %>% group_by(num_cg) %>% summarise(frequency = n())
```
```{r}
biomart_genes %>% distinct(gene_name, num_cg) %>%
  ggplot(aes(x = num_cg)) +
  geom_histogram(bins = 60)
```

```{r}
biomart_genes %>% distinct(gene_name, num_cg) %>% group_by(num_cg) %>% summarise(frequency = n()) %>%
  ggplot(aes(x = num_cg, y = frequency)) +
  geom_point() +
  scale_y_continuous(limits = c(0,500))
```


```{r}
biomart_genes %>% distinct(gene_name, num_cg) %>% group_by(num_cg) %>% summarise(frequency = n()) %>%
  ggplot(aes(x = num_cg, y = frequency)) +
  geom_point()
```
```{r}
go_entrez %>% distinct(go, mean_cg) %>%
  ggplot(aes(x = mean_cg)) +
  geom_histogram()
```


```{r}
wgbs_counts <- readRDS("../output/wgbs_counts.rds")
anno_array <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
anno_array <- anno_array %>% data.frame() %>% 
  mutate(seqnames = chr, start = pos, width = 1, position = paste0(seqnames, "-", start)) %>% 
  relocate(c(seqnames, start, width))
```
```{r}
c(seq_sites = nrow(wgbs_counts), array_sites = nrow(anno_array))
```

# number of CpGs across chromosomes
* seq
```{r}
wgbs_counts %>%
  ggplot(aes(x = seqnames, fill = seqnames)) +
  geom_bar() +
  theme(legend.position = "none")
```

* array
```{r}
anno_array %>%
  ggplot(aes(x = seqnames, fill = seqnames)) +
  geom_bar() +
  theme(legend.position = "none")
```

# overlapping sites

```{r}
overlap <- find_overlaps(as_granges(wgbs_counts), as_granges(anno_array)) %>% data.frame()
nrow(overlap)
```

# Q: How do Array and Sequencing positions relate
* H: Array would be a subset of sequencing
A: Majority overlap

```{r}
list_venn <- list(seq = wgbs_counts$pos, 
                  array = anno_array$position)
ggVennDiagram(list_venn)
```
* is this less that overlap than used to?

# Distance between CpGs
H: CpGs on sequencing will be closer together

```{r}
anno_array <- anno_array %>% .[order(.$seqnames, .$start),] %>%
  group_by(seqnames) %>% mutate(dist_to_next = lead(start) - start) %>% ungroup()
wgbs_counts <- wgbs_counts %>%
  group_by(seqnames) %>% mutate(dist_to_next = lead(start) - start) %>% ungroup()
```


```{r}
anno_array %>%
  ggplot(aes(x = log2(dist_to_next))) +
  geom_density()

anno_array %>%
  ggplot(aes(x = log2(dist_to_next), colour = seqnames)) +
  geom_density()

anno_array %>%
  ggplot(aes(x = seqnames, y = log2(dist_to_next), fill = seqnames)) +
  geom_boxplot() +
  theme(legend.position = "none")
```

```{r}
wgbs_counts %>%
  ggplot(aes(x = log2(dist_to_next))) +
  geom_density()

wgbs_counts %>%
  ggplot(aes(x = log2(dist_to_next), colour = seqnames)) +
  geom_density()

wgbs_counts %>%
  ggplot(aes(x = seqnames, y = log2(dist_to_next), fill = seqnames)) +
  geom_boxplot() +
  theme(legend.position = "none")
```

* Sequencing has smaller gaps

H: CpGs on array overlapping sequencing will have smaller gaps than CpGs that don't
```{r}
anno_array %>% mutate(in_overlap = ifelse(Name %in% overlap$Name, TRUE, FALSE)) %>%
  ggplot(aes(x = log2(dist_to_next), colour = in_overlap)) +
  geom_density() 
```

```{r}
wgbs_counts %>% mutate(in_overlap = ifelse(pos %in% overlap$pos.x, TRUE, FALSE)) %>%
  ggplot(aes(x = log2(dist_to_next), colour = in_overlap)) +
  geom_density()
```


# CpGs and CpG features - Islands etc
```{r}
anno_cpg <- build_annotations(genome="hg19", annotations = "hg19_cpgs")
anno_cpg <- anno_cpg %>% data.frame()

overlap_feature <- find_overlaps(as_granges(wgbs_counts), as_granges(anno_cpg)) %>% data.frame() 
```

* Array
```{r}
rbind(anno_array %>% group_by(Relation_to_Island) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "All"),
  overlap %>% group_by(Relation_to_Island) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "Overlap"),
  anno_array %>% filter(!Name %in% overlap$Name) %>% group_by(Relation_to_Island) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "Array only")) %>%
  ggplot(aes(x = data, y = ratio, fill = Relation_to_Island)) +
  geom_bar(stat = "identity")

rbind(anno_array %>% group_by(Relation_to_Island) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "All"),
  overlap %>% group_by(Relation_to_Island) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "Overlap"),
  anno_array %>% filter(!Name %in% overlap$Name) %>% group_by(Relation_to_Island) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "Array only")) %>%
  ggplot(aes(x = data, y = n, fill = Relation_to_Island)) +
  geom_bar(stat = "identity")
```

```{r}
rbind(anno_array %>% group_by(Relation_to_Island) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "All"),
  overlap %>% group_by(Relation_to_Island) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "Overlap"),
  anno_array %>% filter(!Name %in% overlap$Name) %>% group_by(Relation_to_Island) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "Array only")) %>%
  ggplot(aes(x = data, y = ratio, fill = Relation_to_Island)) +
  geom_bar(stat = "identity", position = "dodge")

rbind(anno_array %>% group_by(Relation_to_Island) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "All"),
  overlap %>% group_by(Relation_to_Island) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "Overlap"),
  anno_array %>% filter(!Name %in% overlap$Name) %>% group_by(Relation_to_Island) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "Array only")) %>%
  ggplot(aes(x = data, y = n, fill = Relation_to_Island)) +
  geom_bar(stat = "identity", position = "dodge")
```
* CpGs only on the array are largely found in the Open Sea (CpG sparse regions)
* they are also a very small proportion of the overall data

* Sequencing

```{r}
rbind(overlap_feature %>% group_by(type) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "All"),
  overlap_feature %>% filter(pos %in% overlap$pos.x) %>% group_by(type) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "Overlap"),
  overlap_feature %>% filter(!pos %in% overlap$pos.x) %>% group_by(type) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "Seq only")) %>%
  ggplot(aes(x = data, y = ratio, fill = type)) +
  geom_bar(stat = "identity")

rbind(overlap_feature %>% group_by(type) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "All"),
  overlap_feature %>% filter(pos %in% overlap$pos.x) %>% group_by(type) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "Overlap"),
  overlap_feature %>% filter(!pos %in% overlap$pos.x) %>% group_by(type) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "Seq only")) %>%
  ggplot(aes(x = data, y = ratio, fill = type)) +
  geom_bar(stat = "identity", position = "dodge")

rbind(overlap_feature %>% group_by(type) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "All"),
  overlap_feature %>% filter(pos %in% overlap$pos.x) %>% group_by(type) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "Overlap"),
  overlap_feature %>% filter(!pos %in% overlap$pos.x) %>% group_by(type) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n), data = "Seq only")) %>%
  ggplot(aes(x = data, y = n, fill = type)) +
  geom_bar(stat = "identity", position = "dodge")
```
* higher proportion of CpGs in islands in array overlap


```{r}
overlap %>% group_by(Relation_to_Island) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n))
```

```{r}
anno_array %>% filter(!Name %in% overlap$Name) %>% group_by(Relation_to_Island) %>% 
  summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n))
```





# Array gene and CpG feature relation
* The majority of CpGs on the array are associated with genes
```{r}
anno_array <- anno_array %>% mutate(has_gene = ifelse(UCSC_RefGene_Name == "", "No", "Yes"))
anno_array %>% group_by(has_gene) %>% summarise(n=n()) %>% ungroup() %>% mutate(ratio = n/sum(n))
```

H: CpGs associated with genes will be more heavily found in CpG Islands
```{r}
anno_array %>% 
  group_by(Relation_to_Island, has_gene) %>% summarise(n=n()) %>%
  ggplot(aes(x = Relation_to_Island, y = n, fill = has_gene)) +
  geom_bar(stat = "identity", position = "fill")

anno_array %>% 
  group_by(Relation_to_Island, has_gene) %>% summarise(n=n()) %>%
  ggplot(aes(x = Relation_to_Island, y = n, fill = has_gene)) +
  geom_bar(stat = "identity", position = "dodge")
```


```{r}
anno_array %>% filter(Name %in% overlap$Name) %>%
  group_by(Relation_to_Island, has_gene) %>% summarise(n=n()) %>%
  ggplot(aes(x = Relation_to_Island, y = n, fill = has_gene)) +
  geom_bar(stat = "identity", position = "fill")

anno_array %>% filter(Name %in% overlap$Name) %>%
  group_by(Relation_to_Island, has_gene) %>% summarise(n=n()) %>%
  ggplot(aes(x = Relation_to_Island, y = n, fill = has_gene)) +
  geom_bar(stat = "identity", position = "dodge")
```

```{r}
anno_array %>% filter(!Name %in% overlap$Name) %>%
  group_by(Relation_to_Island, has_gene) %>% summarise(n=n()) %>%
  ggplot(aes(x = Relation_to_Island, y = n, fill = has_gene)) +
  geom_bar(stat = "identity", position = "fill")

anno_array %>% filter(!Name %in% overlap$Name) %>%
  group_by(Relation_to_Island, has_gene) %>% summarise(n=n()) %>%
  ggplot(aes(x = Relation_to_Island, y = n, fill = has_gene)) +
  geom_bar(stat = "identity", position = "dodge")
```

