---
title: "00_microarrayOntologyAnalysis"
author: "Caitlin Page"
date: "2024-09-24"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: inline
---

Conducting gene set testing of Microarray EPICv1 data using missMethyl.
https://www.bioconductor.org/packages/release/bioc/vignettes/missMethyl/inst/doc/missMethyl.html


## Packages used
```{r}
library(missMethyl)
library(limma)
library(minfi)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)

library(plyranges)
library(dplyr)
library(ggplot2)

library(ggVennDiagram)
```

## Pre-processing
* this the minfi results - not dmrcate

dmrcate_array_anno <- readRDS("../output/dmrcate_array_anno.rds")
nrow(dmrcate_array_anno)
head(dmrcate_array_anno)

dmrcate_array_anno <- readRDS("../output/dmrcate_array_dmr.rds")
nrow(dmrcate_array_anno)
head(dmrcate_array_anno)



dmrcate_array_anno$pos <- paste0(dmrcate_array_anno$seqnames, "-", dmrcate_array_anno$start)
dmrcate_overlap <- find_overlaps(as_granges(dmrcate_array_dmr), as_granges(dmrcate_array_anno)) %>% data.frame()

dmrcate_array_anno <- dmrcate_array_anno %>% mutate(in_dmr = ifelse(pos %in% dmrcate_overlap$pos, TRUE, FALSE))

dmrcate_array_anno <- dmrcate_array_anno %>% mutate(rank_dmr = dmrcate_overlap[match(.$pos.x, dmrcate_overlap$pos), "rank"], rank_dmr = ifelse(is.na(rank_dmr), max(rank_dmr, na.rm = TRUE) + 1, rank_dmr))

dmrcate_array_anno <- find_overlaps(as_granges(dmrcate_array_anno), as_granges(mutate(anno_array, cg_id = rownames(anno_array)))) %>% data.frame()
dmrcate_array_anno <- dmrcate_array_anno %>% mutate(rank_dmr = dmrcate_overlap[match(.$pos.x, dmrcate_overlap$pos), "rank"], rank_dmr = ifelse(is.na(rank_dmr), max(rank_dmr, na.rm = TRUE) + 1, rank_dmr))

dmrcate_array_bio_anno <- annoGeneDmr(anno_array, dmrcate_array_anno, gene_source = "biomart", biomart_genes)
me_gst_dmrcate_array_out <- run_miss_methyl_a(biomart_genes, dmrcate_array_bio_anno)
me_gst_dmrcate_array <- run_miss_methyl_b(run_miss_methyl_a(me_gst_dmrcate_array_out, method = "Wallenius"))

```{r}
DMPs <- readRDS("../output/array_dmps.rds")
head(DMPs)
nrow(DMPs)
```
```{r}
immuneGO <- read.table("../data/gene_sets/GO-immune-system-process.txt", header = FALSE)
names(immuneGO) <- "GOID"
```

```{r}
topCpGs <- DMPs[1:10000,]
sigCpGs <- rownames(topCpGs)
sigCpGs[1:10]
```

```{r}
# Check number of genes that significant CpGs are annotated to
check <- getMappedEntrezIDs(sig.cpg = sigCpGs)
```

```{r}
length(check$sig.eg)
```

## Bias corrected GST

```{r}
gst <- gometh(sig.cpg=sigCpGs, all.cpg=rownames(DMPs), collection="GO",
              plot.bias=TRUE)
```
```{r}
system.time(gometh(rownames(DMPs[1:10000,]), all.cpg = rownames(DMPs), collection = "GO", plot.bias = TRUE))
```

```{r}
sig_terms_array <- topGSA(gst, n = 1000) %>% filter(FDR < 0.05) %>% .[order(.$FDR),]
sig_terms_array$GOID <- rownames(sig_terms_array)
sig_terms_array$rank <- 1:nrow(sig_terms_array)
nrow(sig_terms_array)
head(sig_terms_array)
#saveRDS(sig_terms_array, "../output/mm_array_res.rds")
```


```{r}
sig_terms_array[1:10,] %>% mutate(in_imm = ifelse(GOID %in% immuneGO$GOID, TRUE, FALSE)) %>% .[order(.$FDR, decreasing = TRUE),] %>%
  ggplot(aes(x = -log10(FDR), y = factor(.data$TERM, levels = .data$TERM), colour = in_imm, size = N)) +
  geom_point() +
  geom_vline(xintercept = 0.05, type = "dotted") 
```


## Biased GST

```{r}
gst.biased <- gometh(sig.cpg=sigCpGs, all.cpg=rownames(DMPs), collection="GO",
              prior.prob=FALSE)

sig_terms_array_biased <- topGSA(gst.biased, n=10000) %>% filter(FDR < 0.05) %>% .[order(.$FDR),]
sig_terms_array_biased$GOID <- rownames(sig_terms_array_biased)
sig_terms_array_biased$rank <- 1:nrow(sig_terms_array_biased)
nrow(sig_terms_array_biased)
head(sig_terms_array_biased)
#saveRDS(sig_terms_array_biased, "../output/mm_array_null.rds")
```

## compare bias and corrected

```{r}
list_venn <- list(correct = sig_terms_array$TERM, bias = sig_terms_array_biased$TERM)
ggVennDiagram(list_venn)
```

```{r}
truth_gst_100_array <- rbind(
  sig_terms_array[1:100,] %>% mutate(in_imm = cumsum(rownames(.) %in% immuneGO$GOID), 
                                     method = "Bias corrected") %>% 
    .[,colnames(.) %in% c("GOID", "rank", "in_imm", "method")],
  sig_terms_array_biased[1:100,] %>% mutate(in_imm = cumsum(rownames(.) %in% immuneGO$GOID), method = "Biased") %>% 
    .[,colnames(.) %in% c("GOID", "rank", "in_imm", "method")])

truth_gst_100_array %>%
  ggplot(aes(x = rank, y = in_imm, colour = method)) +
  geom_line() +
  labs(x = "Rank", y = "Cumulative no. in immune sets")
```


# promoters

```{r}
gst_promoter <- gometh(sig.cpg=sigCpGs, all.cpg=rownames(DMPs), collection="GO",
              plot.bias=TRUE, genomic.features = "TSS1500")
```
```{r}
gst_promoter
sig_terms_array_promoter <- gst_promoter %>% .[order(.$FDR),] %>% filter(FDR < 0.05)
nrow(sig_terms_array_promoter)
head(sig_terms_array_promoter)
```

```{r}
gst_promoter.biased <- gometh(sig.cpg=sigCpGs, all.cpg=rownames(DMPs), collection="GO",
              prior.prob = FALSE, genomic.features = "TSS1500")
```
```{r}
sig_terms_array_promoter_biased <- gst_promoter.biased %>% .[order(.$FDR),] %>% filter(FDR < 0.05)
nrow(sig_terms_array_promoter_biased)
head(sig_terms_array_promoter_biased)
```

```{r}
list_venn <- list(correct = sig_terms_array_promoter$TERM, bias = sig_terms_array_promoter_biased$TERM)
ggVennDiagram(list_venn)
```



## Compare gene and promoter

* Bias corrected
```{r}
list_venn <- list(gene = sig_terms_array$TERM, prom = sig_terms_array_promoter$TERM)
ggVennDiagram(list_venn)
```

* Biased
```{r}
list_venn <- list(gene = sig_terms_array_biased$TERM, prom = sig_terms_array_promoter_biased$TERM)
ggVennDiagram(list_venn)
```



## with kegg
```{r}
gst.kegg <- gometh(sig.cpg=sigCpGs, all.cpg=rownames(DMPs), collection="KEGG")
## All input CpGs are used for testing.
topGSA(gst.kegg, n=10)

```


