---
title: "00_obtainGenes"
author: "Caitlin Page"
date: "2024-09-10"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: inline
---

```{r}
sourceGenes <- function(counts, gene_source = c("biomaRt", "ExperimentHub"), gene_feature=c("gene", "promoter")) {
  gene_source <- match.arg(tolower(gene_source), c("biomart", "experimenthub"))
  gene_feature <- match.arg(tolower(gene_feature), c("gene", "promoter"))

  if(gene_source == "biomart") {
    ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl", version = "GRCh37")
    genes <- getBM(
      attributes = c("ensembl_gene_id", "external_gene_name", "ensembl_transcript_id",
                     "chromosome_name", "start_position", "end_position", "strand",
                     "transcription_start_site" , "transcript_start", "transcript_end", "transcript_length"),
      filters = "chromosome_name",
      values = gsub("chr", "", unique(counts$seqnames)[1:22]),
      mart = ensembl
    )
    genes <- genes %>% .[order(.$chromosome_name, .$start_position), ]
    colnames(genes) <- c("ensembl_gene_id", "gene_name", "ensembl_transcript_id", "seqnames", "start", "end", "strand", "TSS", "t_start", "t_end", "t_length")
    genes <- genes %>% group_by(ensembl_gene_id) %>% filter(t_length == max(t_length)) %>% ungroup()
    genes$seqnames <- paste0("chr", as.character(genes$seqnames))
    genes <- genes %>% group_by(ensembl_gene_id) %>% mutate(num_t = 1:dplyr::n()) %>% filter(num_t == 1) %>% ungroup()
  } else if (gene_source == "experimenthub") {
    eh <- ExperimentHub()
    genes <- eh[["EH3132"]]
  }
  genes <- genes %>% data.frame() %>% ungroup()
  # ensure ensembl_id name consistency
  colnames(genes)[grepl("ENSG", genes[1,])] <- "ensembl_gene_id"
  # double check only 1 ensembl id per gene entry
  genes$width <- genes$end - genes$start + 1
  genes <- genes %>% group_by(ensembl_gene_id) %>% filter(width == max(width)) %>% ungroup()
  # add random identifier
  genes$gene_num <- 1:nrow(genes)
  #is it promoter???
  if(length(gene_feature) == 1 & gene_feature == "promoter") {
    genes$start <- ifelse(genes$strand == 1, genes$start - 2000, genes$end - 500)
    genes$end <- ifelse(genes$strand == 1, genes$start + 2500, genes$end + 2000)
    genes <- genes[,!colnames(genes) == "width"]
  }
  # add number of cpgs overlapping gene
  gene_cpg <- find_overlaps(as_granges(genes), as_granges(counts)) %>% data.frame()
  gene_cpg <- gene_cpg %>% group_by(gene_num) %>% summarise(n_cpg = dplyr::n()) %>% ungroup() %>% data.frame()

  genes$num_cg <- gene_cpg[match(genes$gene_num, gene_cpg$gene_num), "n_cpg"]
  genes[is.na(genes)] <- 0
  genes
}

annoGeneDmr <- function(counts, dmrs, gene_source = c("biomart", "experimenthub", "self"), genes) { #why do I have the gene source here?? why???
  gene_source <- match.arg(tolower(gene_source), c("biomart", "experimenthub", "self"))
  genes <- data.frame(genes)
  # ensure ensembl_id name consistency
  colnames(genes)[grepl("ENSG", genes[1,])] <- "ensembl_gene_id"
  # double check only 1 ensembl id per gene entry
  genes$width <- genes$end - genes$start + 1
  genes <- genes %>% group_by(ensembl_gene_id) %>% filter(width == max(width)) %>% ungroup()
  # add random identifier
  genes$gene_num <- 1:nrow(genes)
  # add number of cpgs overlapping gene
  gene_cpg <- find_overlaps(as_granges(genes), as_granges(counts)) %>% data.frame()
  gene_cpg <- gene_cpg %>% group_by(gene_num) %>% summarise(n_cpg = n()) %>% ungroup() %>% data.frame()

  genes$num_cg <- gene_cpg[match(genes$gene_num, gene_cpg$gene_num), "n_cpg"]
  genes[is.na(genes)] <- 0
  if(!"TSS" %in% colnames(genes)) {
    genes$TSS <- ifelse(genes$strand == "+", genes$start, genes$end) # above I have strand for 1/-1
  }
  # assume dmrs are already in order of most to least signif
  dmrs$rank <- 1:nrow(dmrs)
  # add if gene overlaps dmr
  overlap_dmr <- find_overlaps(as_granges(dmrs), as_granges(genes)) %>% data.frame() %>%
    mutate(combo = paste0(ensembl_gene_id, "-DMR", rank))
  anno_dmr <- pair_nearest(as_granges(dmrs), as_granges(genes)) %>% data.frame() %>%
    mutate(combo = paste0(ensembl_gene_id, "-DMR", rank))
  #dist gene to dmr
  anno_dmr <- anno_dmr %>%
    mutate(tss_start = TSS - granges.x.start, tss_end = TSS - granges.x.end) %>%
    mutate(tss_rel_peak = ifelse(combo %in% overlap_dmr$combo, "Overlap", NA),
            tss_rel_peak = case_when(is.na(tss_rel_peak) & tss_start > 0 & tss_end > 0 & tss_end < tss_start ~ "Peak_upstream",
                                     is.na(tss_rel_peak) & tss_start < 0 & tss_end < 0 &  tss_end < tss_start ~ "Peak_downstream", TRUE ~ tss_rel_peak),
           tss_rel_peak = case_when(tss_rel_peak == "Peak_upstream" & granges.y.strand == "-" ~ "Peak_downstream",
                                   tss_rel_peak == "Peak_downstream" & granges.y.strand == "-" ~ "Peak_upstream", TRUE ~ tss_rel_peak),
          min_dist = case_when(tss_rel_peak == "Overlap" ~ 0, tss_rel_peak == "Peak_downstream" & granges.y.strand == "+" ~ tss_start,
                                tss_rel_peak == "Peak_downstream" & granges.y.strand == "-" ~ tss_end, tss_rel_peak == "Peak_upstream" & granges.y.strand == "+" ~ tss_end,
                                tss_rel_peak == "Peak_upstream" & granges.y.strand == "-" ~ tss_start))
  anno_dmr

}
```




## Introduction

```{r}
library(plyranges)
library(dplyr)
library(ggplot2)
```

```{r}
library(AnnotationHub)
ah <- AnnotationHub()
```
```{r}
ah$genome
query(query(query(ah, "hg19"), "Homo sapiens"), "Ensembl genes")
```
```{r}
ah_a <- ah[["AH5046"]]
ah_a[1:5,] %>% data.frame()
```
```{r}
query(query(query(ah, "hg19"), "Homo sapiens"), "UCSC genes")
ah[["AH5036"]][1:5,] %>% data.frame()
```
- where is the TSS hiding??
- dont want to have to need biomart

```{r}
#query(query(query(query(ah, "hg19"), "Homo sapiens"), "GRanges"), "GENCODE")
```
- ensembl, ucsc, and gencode don't give tss coords
```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
TxDb.Hsapiens.UCSC.hg19.knownGene
```
```{r}
genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
```
- maybe like a combination of trancripts and genes?
- but that seems terrible


```{r}
eh <- ExperimentHub()
eh_genes <- eh[["EH3132"]]
eh_genes
```
```{r}
eh["EH3132"]
```

```{r}
eh$
head(unique(eh$species))
```
```{r}
query(eh, "Homo sapiens")
```
- use biomart for now
```{r}
library(biomaRt)
listEnsemblArchives()
```
```{r}
ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl", version = "GRCh37")
```
- need hg19 which is grch37
```{r}
listAttributes(ensembl)
```
- the old ensembl website has canonical transcript stuff
- so why can't i find it here
- this is annoying

- but ensembl transcript is longest, most high exp, and matches other databases
- maybe run stuff just picking a random for now lol
```{r}
wgbs_counts <- readRDS("../output/wgbs_counts.rds")
```

```{r}
biomart_genes <- getBM(
        attributes = c("ensembl_gene_id", "external_gene_name", "ensembl_transcript_id",
            "chromosome_name", "start_position", "end_position", "strand",
            "transcription_start_site" , "transcript_start", "transcript_end", "transcript_length"),
        filters = "chromosome_name",
        values = gsub("chr", "", unique(wgbs_counts$seqnames)[1:22]),
        mart = ensembl
    )
biomart_genes <- biomart_genes %>% .[order(.$chromosome_name, .$start_position), ]
colnames(biomart_genes) <- c("ensembl_gene_id", "gene_name", "ensembl_transcript_id", "seqnames", "start", "end", "strand", "TSS", "t_start", "t_end", "t_length")
biomart_genes
```
```{r}
biomart_genes %>% mutate(check_t_length = t_end - t_start, same = ifelse(t_length == check_t_length, TRUE, FALSE))  %>% group_by(same) %>% summarise(n())
biomart_genes %>% mutate(check_t_length = t_end - t_start, same = ifelse(t_length == check_t_length, TRUE, FALSE)) %>% filter(same == TRUE)
```
- so the ensembl transcript length isn't taken from the start and stop
- must be because it includes UTR's?
- but does it change the largest transcript per gene??
```{r}
biomart_genes %>% mutate(check_t_length = t_end - t_start, same = ifelse(t_length == check_t_length, TRUE, FALSE), diff = check_t_length - t_length) %>% group_by(ensembl_gene_id) %>% mutate(max_trans = max(t_length), max_check = max(check_t_length), same_max = ifelse(max_trans == t_length & max_check == check_t_length, TRUE, FALSE))
```
- ok so that doesn't really match
- maybe just go with the transcript length one?

```{r}
biomart_genes <- biomart_genes %>% group_by(ensembl_gene_id) %>% filter(t_length == max(t_length)) %>% ungroup() %>% mutate(seqnames = as.character(seqnames), seqnames = paste0("chr", seqnames))
biomart_genes
```
```{r}
biomart_genes %>% distinct(ensembl_gene_id) %>% nrow()
```

```{r}
biomart_genes <- biomart_genes %>% group_by(ensembl_gene_id) %>% mutate(num_t = 1:n()) %>% filter(num_t == 1) %>% ungroup() 
biomart_genes
```
- hold up these tss's are just matching the start site
- that's not how it's supposed to work
- hold up they used to be different - but seems like by picking the longest transcript - (or trying to), the tss is now same as the start
```{r}
biomart_genes %>% group_by(gene_name) %>% summarise(n=n()) %>% filter(n>1)
```
?????
- let's ignore that
```{r}
biomart_genes %>% group_by(gene_name) %>% mutate(n=n()) %>% filter(n>1) %>% .[order(.$gene_name),]
```
- ok so now bit nervous a screw up has happened because we have different ensembl-id but same gene-name
- and lots of these are different chrom
- i think i want to check on these before i do tss stuff
```{r}
biomart_cg <- find_overlaps(as_granges(biomart_genes), as_granges(wgbs_counts)) %>% data.frame() %>% group_by(ensembl_gene_id) %>% summarise(num_cg = n())
biomart_cg <- data.frame(biomart_cg)
```

```{r}
biomart_genes <- biomart_genes %>% mutate(num_cg = biomart_cg[match(.$ensembl_gene_id, biomart_cg$ensembl_gene_id), "num_cg"],
                                          num_cg = ifelse(is.na(num_cg), 0, num_cg))
```


```{r}
nrow(eh_genes)
nrow(distinct(eh_genes, gene))
nrow(distinct(eh_genes, symbol))
```
```{r}
eh_genes %>% group_by(symbol) %>% mutate(n=n()) %>% filter(n>1) %>% .[order(.$symbol),]
```
```{r}
(eh_genes %>% group_by(symbol) %>% mutate(n=n()) %>% filter(n>1) %>% ungroup() %>% .$symbol %>% unique(.)) %in% (biomart_genes %>% group_by(gene_name) %>% mutate(n=n()) %>% filter(n>1) %>% .$gene_name %>% unique(.)) %>% summary()
(eh_genes %>% group_by(symbol) %>% mutate(n=n()) %>% filter(n>1) %>% ungroup() %>% .$symbol %>% unique(.)) %in% (biomart_genes %>% group_by(gene_name) %>% mutate(n=n()) %>% filter(n>1) %>% .$gene_name %>% unique(.)) 
```
```{r}
(biomart_genes %>% group_by(gene_name) %>% mutate(n=n()) %>% filter(n>1) %>% .$gene_name %>% unique(.)) %in% (eh_genes %>% group_by(symbol) %>% mutate(n=n()) %>% filter(n>1) %>% ungroup() %>% .$symbol %>% unique(.)) %>% summary()
(biomart_genes %>% group_by(gene_name) %>% mutate(n=n()) %>% filter(n>1) %>% .$gene_name %>% unique(.)) %in% (eh_genes %>% group_by(symbol) %>% mutate(n=n()) %>% filter(n>1) %>% ungroup() %>% .$symbol %>% unique(.))
```
- so all biomart in eh_genes
- but eh_genes has some extras
```{r}
biomart_genes %>% group_by(gene_name) %>% summarise(n=n()) %>% .[order(.$n, decreasing = TRUE),]
```
```{r}
biomart_genes <- biomart_genes %>% group_by(gene_name) %>% mutate(num_gene_name = n()) %>% ungroup() %>% filter(num_gene_name == 1) %>% data.frame()
biomart_genes
```




# found a paper that used genomicfeatures and txdb to get start sites
- genomfeatures vignette - promoters fn is area around tss for transcript
```{r}
library(GenomicFeatures)
promoters(TxDb.Hsapiens.UCSC.hg19.knownGene, upstream = 0, downstream = 0)
```
- except that's way more than the genes - and no way to match unless overlapping i guess
```{r}
genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene)
```
- stick with biomart genes for now i think

# biomart
```{r}
biomart_genes
```
# allow for multiple sources with diff col names
- which col has the ensembl_gene_id?? (ENSG)
```{r}
biomart_genes
```

```{r}
colnames(biomart_genes)[grepl("ENSG", biomart_genes[1,])]
```
- i want the index of the true
```{r}
biomart_genes2 <- biomart_genes[1:5,]
biomart_genes2
colnames(biomart_genes2)[grepl("ENSG", biomart_genes2[1,])] <- "ensembl_id"
biomart_genes2
```
```{r}
rm(biomart_genes2)
```

