---
title: "00_obtainGenes"
author: "Caitlin Page"
date: "2024-09-10"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: inline
---


## Packages

```{r}
library(plyranges)
library(dplyr)
library(ggplot2)

library(AnnotationHub)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(GenomicFeatures)
library(ExperimentHub)
library(biomaRt)
```

## Description

A description of some Bioconductor annotation resources: [here](https://bioconductor.github.io/BiocWorkshops/introduction-to-bioconductor-annotation-resources.html)

# Possible methods to obtain Genes
* AnnotationHub
* ExperimentHub
* TxDb object
* biomaRt

We want flexibility to be inbuilt in our function, and acknowledge that individuals will have different preferences for their gene annotations.
Information that we will need to be flexible with:
* species
* genome version

Minimum information we will require from the resource
* species
* genome version

## Using [AnnotationHub](https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html)
* What: access web resource
This connects to different files (eg VCF, bed, wig) from various locations (UCSC, Ensembl)
```{r}
ah <- AnnotationHub()
ah
```
* access information via a query with what you are interested in
```{r}
query(ah, c("hg19", "Homo sapiens"))
```

```{r}
query(ah, c("hg19", "Homo sapiens", "Ensembl genes"))
```
* can then retrieve the information
```{r}
ah_a <- ah[["AH5046"]]
ah_a %>% data.frame()
```
```{r}
query(ah, c("hg19", "Homo sapiens", "UCSC genes"))
ah[["AH5036"]][1:5,] %>% data.frame()
```
- where is the TSS hiding??
- dont want to have to need biomart
```{r}
query(ah, c("hg19", "Homo sapiens"))
```


# Summary
* A web resource so it is potentially unstable
* Provides access to the widest range of data sources
* ensembl, ucsc, and gencode don't give tss coords

# Considerations for inclusion
* the query used
* package could be a recommends

## Using TxDb object
* These are packages installed for the species and genome version interested in
```{r}
TxDb.Hsapiens.UCSC.hg19.knownGene
```
```{r}
genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
```

# Summary
* Uses Entrez Gene Ids (also used for GST)
* Can access genes and transcripts

# Considerations for inclusion
* Explicit documentation that the selected package must be installed separately
* Do we care about the transcripts?
* No gene symbol names - would need to be added
* No TSS information

## Using Experiment Hub

```{r}
eh <- ExperimentHub()
eh_genes <- eh[["EH3132"]]
eh_genes
```
```{r}
eh["EH3132"]
```

```{r}
head(unique(eh$species))
```
```{r}
query(eh, "Homo sapiens")
```

# Summary
* It's genes from DMRCate 
* As this is an experiment, not appropriate 


## Using biomart

```{r eval=FALSE}
listEnsemblArchives()
```
```{r eval=FALSE}
ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl", version = "GRCh37")
```
- need hg19 which is grch37
```{r eval=FALSE}
listAttributes(ensembl)
```
- the old ensembl website has canonical transcript stuff
- so why can't i find it here
- this is annoying

- but ensembl transcript is longest, most high exp, and matches other databases
- maybe run stuff just picking a random for now lol
```{r eval=FALSE}
wgbs_counts <- readRDS("../output/wgbs_counts.rds")
```

```{r eval=FALSE}
biomart_genes <- getBM(
        attributes = c("ensembl_gene_id", "external_gene_name", "ensembl_transcript_id",
            "chromosome_name", "start_position", "end_position", "strand",
            "transcription_start_site" , "transcript_start", "transcript_end", "transcript_length"),
        filters = "chromosome_name",
        values = gsub("chr", "", unique(wgbs_counts$seqnames)[1:22]),
        mart = ensembl
    )
biomart_genes <- biomart_genes %>% .[order(.$chromosome_name, .$start_position), ]
colnames(biomart_genes) <- c("ensembl_gene_id", "gene_name", "ensembl_transcript_id", "seqnames", "start", "end", "strand", "TSS", "t_start", "t_end", "t_length")
head(biomart_genes)
```

# Summary
* Uses a web resource so potentially unstable
* Large depth in diversity of information

# Considerations for inclusion
* the way species and genome version are defined in the ensembl query
* what type of information should be taken from it

## Developing the function

# Features
* Allow for multiple potential inputs: biomaRt and TxDb
* Contains TSS or estimates TSS (start site with strand information)
* Identifies gene or promoter regions
* Promoter defined as [TSS-2000, TSS + 500]
* The number of CpGs: number of overall CpGs (not tested CpGs - this is experiment specific)

```{r}
sourceGenes <- function(counts, gene_source = c("biomaRt", "ExperimentHub"), gene_feature=c("gene", "promoter")) {
  gene_source <- match.arg(tolower(gene_source), c("biomart", "experimenthub"))
  gene_feature <- match.arg(tolower(gene_feature), c("gene", "promoter"))

  if(gene_source == "biomart") {
    ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl", version = "GRCh37")
    genes <- getBM(
      attributes = c("ensembl_gene_id", "external_gene_name", "ensembl_transcript_id",
                     "chromosome_name", "start_position", "end_position", "strand",
                     "transcription_start_site" , "transcript_start", "transcript_end", "transcript_length"),
      filters = "chromosome_name",
      values = gsub("chr", "", unique(counts$seqnames)[1:22]),
      mart = ensembl
    )
    genes <- genes %>% .[order(.$chromosome_name, .$start_position), ]
    colnames(genes) <- c("ensembl_gene_id", "gene_name", "ensembl_transcript_id", "seqnames", "start", "end", "strand", "TSS", "t_start", "t_end", "t_length")
    genes <- genes %>% group_by(ensembl_gene_id) %>% filter(t_length == max(t_length)) %>% ungroup()
    genes$seqnames <- paste0("chr", as.character(genes$seqnames))
    genes <- genes %>% group_by(ensembl_gene_id) %>% mutate(num_t = 1:dplyr::n()) %>% filter(num_t == 1) %>% ungroup()
  } else if (gene_source == "experimenthub") {
    eh <- ExperimentHub()
    genes <- eh[["EH3132"]]
  }
  genes <- genes %>% data.frame() %>% ungroup()
  # ensure ensembl_id name consistency
  colnames(genes)[grepl("ENSG", genes[1,])] <- "ensembl_gene_id"
  # double check only 1 ensembl id per gene entry
  genes$width <- genes$end - genes$start + 1
  genes <- genes %>% group_by(ensembl_gene_id) %>% filter(width == max(width)) %>% ungroup()
  # add random identifier
  genes$gene_num <- 1:nrow(genes)
  #is it promoter???
  if(length(gene_feature) == 1 & gene_feature == "promoter") {
    genes$start <- ifelse(genes$strand == 1, genes$start - 2000, genes$end - 500)
    genes$end <- ifelse(genes$strand == 1, genes$start + 2500, genes$end + 2000)
    genes <- genes[,!colnames(genes) == "width"]
  }
  # add number of cpgs overlapping gene
  gene_cpg <- find_overlaps(as_granges(genes), as_granges(counts)) %>% data.frame()
  gene_cpg <- gene_cpg %>% group_by(gene_num) %>% summarise(n_cpg = dplyr::n()) %>% ungroup() %>% data.frame()

  genes$num_cg <- gene_cpg[match(genes$gene_num, gene_cpg$gene_num), "n_cpg"]
  genes[is.na(genes)] <- 0
  genes
}
```



