---
title: "02_cpgAvailVsTested"
author: "Caitlin Page"
date: "2025-03-04"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: inline
---

## Introduction

Not all of the available CpGs will be utilised in the DM testing procedure - some will be filtered out because they have no/low counts.

[Wijesooria et al., 2022](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009935) highlights the importance of using a correct background gene set in regards to gene set testing analysis. This means using the only the genes that were testing in the differential analysis, as opposed to all possible genes.

This will be relevant to us in both the genes input as the background - making sure that they are only included if they contain CpGs that were tested, and in the bias correction - making sure that only CpGs used in the differential testing are provided as the bias.

In my early analysis, I did not realise this was an important feature, and had been getting my bias (number of CpGs) by merely overlapping the CpGs to a gene.
Here, I wish to do some analysis to understand how using only the tested CpGs changes the results, and the relationships between them all.


## Packages

```{r}
library(plyranges)
library(dplyr)
library(ggplot2)
library(BSDA)
library(ggVennDiagram)
library(gplots)

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
library(IlluminaHumanMethylationEPICmanifest)

library(annotatr)
```

## Pre-processed files
```{r}
wgbs_counts <- readRDS("../output/wgbs_counts.rds")
anno_array <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
anno_array <- anno_array %>% data.frame() %>% 
  mutate(seqnames = chr, start = pos, width = 1,
         position = paste0(chr, "-", pos)) %>% relocate(c(seqnames, start, width))

DMPs <- readRDS("../output/array_dmps.rds") 
DMPs$pos <- paste0(DMPs$seqnames, "-", DMPs$start)
dmrcate_array_anno <- readRDS("../output/dmrcate_array_anno.rds")
dmrcate_array_anno <- dmrcate_array_anno@ranges %>% data.frame() %>%
  mutate(pos = paste0(seqnames, "-", start))
dmrcate_array_anno <- dmrcate_array_anno %>% 
  mutate(cg_id = anno_array[match(.$pos, anno_array$position), "Name"])

bsseq_res <- readRDS("../output/bsseq_res.rds")
dmrcate_seq_anno <- readRDS("../output/dmrcate_seq_anno.rds")
dmrcate_seq_anno <- dmrcate_seq_anno@ranges %>% data.frame() %>%
  mutate(pos = paste0(seqnames, "-", start))

biomart_genes <- readRDS("../output/genes_biomart.rds")
```



# overlap
* Seq
```{r}
list_venn <- list(seq = wgbs_counts$pos, 
                  bsseq = bsseq_res$position,
                  dmrcate = dmrcate_seq_anno$pos)
ggVennDiagram(list_venn)
```

* array
```{r}
list_venn <- list(seq = anno_array$position, 
                  minfi = DMPs$pos,
                  dmrcate = dmrcate_array_anno$pos)
ggVennDiagram(list_venn)
```


## Compare distance between tested sites
Q: Are there significant differences (on average) for the distance between sites depending on if they were tested or not?

```{r}
dmrcate_seq_anno <- dmrcate_seq_anno %>% .[order(.$seqnames, .$start),] %>%
  group_by(seqnames) %>% mutate(dist_to_next = lead(start) - start) %>% ungroup()
bsseq_res <- bsseq_res %>% .[order(.$seqnames, .$start),] %>%
  group_by(seqnames) %>% mutate(dist_to_next = lead(start) - start) %>% ungroup()

dmrcate_array_anno <- dmrcate_array_anno %>% .[order(.$seqnames, .$start),] %>%
  group_by(seqnames) %>% mutate(dist_to_next = lead(start) - start) %>% ungroup()
DMPs <- DMPs %>% .[order(.$seqnames, .$start),] %>%
  group_by(seqnames) %>% mutate(dist_to_next = lead(start) - start) %>% ungroup()
```

# WGBS

```{r}
dmrcate_seq_anno %>%
  ggplot(aes(x = log2(dist_to_next), colour = seqnames)) +
  geom_density()

dmrcate_seq_anno %>%
  ggplot(aes(y = log2(dist_to_next), x = seqnames, colour = seqnames)) +
  geom_boxplot() +
  theme(legend.position = "none")
```

```{r}
bsseq_res %>%
  ggplot(aes(x = log2(dist_to_next), colour = seqnames)) +
  geom_density()

bsseq_res %>%
  ggplot(aes(y = log2(dist_to_next), x = seqnames, colour = seqnames)) +
  geom_boxplot() +
  theme(legend.position = "none")
```

```{r}
t.test(dmrcate_seq_anno$dist_to_next, bsseq_res$dist_to_next)
```

# compare all sites to tested sites

```{r}
wgbs_counts <- wgbs_counts %>% group_by(seqnames) %>% mutate(dist_to_next = lead(start) - start) %>% ungroup()
c(mean = mean(wgbs_counts$dist_to_next, na.rm=TRUE), sd = sd(wgbs_counts$dist_to_next, na.rm=TRUE))
```
```{r}
z.test(dmrcate_seq_anno$dist_to_next, mu=mean(wgbs_counts$dist_to_next, na.rm=TRUE), sigma.x = sd(wgbs_counts$dist_to_next, na.rm=TRUE))
```
* do not reject null
* basically the tested sites are not a distinct group in terms of the distance between them

# array

```{r}
dmrcate_array_anno %>%
  ggplot(aes(x = log2(dist_to_next), colour = seqnames)) +
  geom_density()

dmrcate_array_anno %>%
  ggplot(aes(y = log2(dist_to_next), x = seqnames, colour = seqnames)) +
  geom_boxplot() +
  theme(legend.position = "none")
```

```{r}
DMPs %>%
  ggplot(aes(x = log2(dist_to_next), colour = seqnames)) +
  geom_density()

DMPs %>%
  ggplot(aes(y = log2(dist_to_next), x = seqnames, colour = seqnames)) +
  geom_boxplot() +
  theme(legend.position = "none")
```

```{r}
t.test(dmrcate_array_anno$dist_to_next, DMPs$dist_to_next)
```
* 2 groups are not distinct as expected

# Compare tested and all
```{r}
anno_array <- anno_array %>% .[order(.$seqnames, .$start),] %>%
  group_by(seqnames) %>% mutate(dist_to_next = lead(start) - start) %>% ungroup()
c(mean = mean(anno_array$dist_to_next, na.rm = TRUE), sd = sd(anno_array$dist_to_next, na.rm = TRUE))
```

```{r}
z.test(dmrcate_array_anno$dist_to_next, mu = mean(anno_array$dist_to_next, na.rm = TRUE), sigma.x = sd(anno_array$dist_to_next, na.rm = TRUE))
```
* in the array, tested sites are a significant separate group to overall in regards to the distance between the sites

# Compare array and seq

```{r}
t.test(dmrcate_seq_anno$dist_to_next[!is.na(dmrcate_seq_anno$dist_to_next)], dmrcate_array_anno$dist_to_next[!is.na(dmrcate_array_anno$dist_to_next)])
```
* 2 groups are significantly different
* sequencing has much smaller gaps between tested CpGs


## Distribution of tested CpGs among CpG annotations - islands etc

# array
* relationship to CpG annotation is included in the EPIC annotation file/package
```{r}
anno_array <- data.frame(anno_array)
head(anno_array)
```

```{r}
dmrcate_array_anno <- data.frame(dmrcate_array_anno)
dmrcate_array_anno$cpg_feature <- anno_array[match(dmrcate_array_anno$cg_id, anno_array$Name), "Relation_to_Island"]
```

```{r}
dmrcate_array_anno %>% group_by(cpg_feature) %>% summarise(n=n()) %>% ungroup() %>% mutate(total = sum(n), prop = n/sum(n))
```

```{r}
dmrcate_array_anno %>%
  ggplot(aes(x = cpg_feature, fill = cpg_feature)) +
  geom_bar() +
  theme(legend.position = "none")
```
* so far I've just been comparing array and seq
* not diff between tested and non tested
* i still think some kind of chi-square - because I want to compare how the proportions change/distributions change
* dist to next would be a z test I think - because the tested is a subgroup of the all

```{r}
dt <- cbind(dmrcate_array_anno %>% group_by(cpg_feature) %>% summarise(tested_n=n()), 
      all_n = anno_array %>% group_by(Relation_to_Island) %>% summarise(all_n=n()) %>% .$all_n)
rownames(dt) <- dt$cpg_feature
dt <- dt[,2:3]
dt
```
```{r}
balloonplot(t(as.table(as.matrix(dt))),
      main = "Array CpG features", xlab = "", ylab = "", label = FALSE, show.margins = FALSE)
```
* frequencies appear to be very similar

# sequencing

```{r}
anno_cpg <- build_annotations(genome="hg19", annotations = "hg19_cpgs")
anno_cpg <- anno_cpg %>% data.frame()

overlap_feature <- find_overlaps(as_granges(wgbs_counts), as_granges(anno_cpg)) %>% data.frame() 
dmrcate_seq_anno <- find_overlaps(as_granges(dmrcate_seq_anno), as_granges(anno_cpg)) %>% data.frame()
```


```{r}
dt <- cbind(dmrcate_seq_anno %>% group_by(type) %>% summarise(tested_n=n()), 
      all_n = overlap_feature %>% group_by(type) %>% summarise(all_n=n()) %>% .$all_n)
rownames(dt) <- dt$type
dt <- dt[,2:3]
dt
```
```{r}
balloonplot(t(as.table(as.matrix(dt))),
      main = "Seq CpG features", xlab = "", ylab = "", label = FALSE, show.margins = FALSE)
```
* similar frequencies

# Seq and Array
* For Array, OpenSea is Inter, and combine N and S shelves and shores
```{r}
dt <- cbind(dmrcate_array_anno %>% group_by(cpg_feature) %>% summarise(array_test=n()), 
      array_all = anno_array %>% group_by(Relation_to_Island) %>% summarise(all_n=n()) %>% .$all_n)
dt <- rbind(dt, c(cpg_feature = "shelf", array_test = dt[2,2] + dt[5,2], array_all = dt[2,3] + dt[5,3]),
      c(cpg_feature = "shore", array_test = dt[3,2] + dt[6,2], array_all = dt[3,3] + dt[6,3])) %>% .[c(4,1,7,8),]
rownames(dt) <- NULL
dt <- cbind(cbind(dmrcate_seq_anno %>% group_by(type) %>% summarise(seq_test=n()), 
      seq_all = overlap_feature %>% group_by(type) %>% summarise(all_n=n()) %>% .$all_n), dt)
rownames(dt) <- dt$type
dt <- dt[,c(2,3,5,6)]
dt
```

```{r}
balloonplot(t(as.table(as.matrix(dt))),
      main = "CpG features", xlab = "", ylab = "", label = FALSE, show.margins = FALSE)
```


```{r}
dmrcate_array_anno %>%
  ggplot(aes(x = log2(dist_to_next), colour = cpg_feature)) +
  geom_density()
anno_array %>%
  ggplot(aes(x = log2(dist_to_next), colour = Relation_to_Island)) +
  geom_density()
```

```{r}
dmrcate_seq_anno %>%
  ggplot(aes(x = log2(dist_to_next), colour = type)) +
  geom_density()

overlap_feature %>%
  ggplot(aes(x = log2(dist_to_next), colour = type)) +
  geom_density()
```
* CpGs in islands have a shorter distance between them




