---
title: "02_exploreBiasGenes"
author: "Caitlin Page"
date: "2024-09-05"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: inline
---

## Introduction

```{r}
library(plyranges)
library(dplyr)
library(reshape2)
library(ggplot2)
library(ExperimentHub)
library(BSgenome.Hsapiens.UCSC.hg19)
```

```{r}
wgbs_counts <- readRDS("../output/wgbs_counts.rds")
bsseq_res <- readRDS("../output/bsseq_res.rds")
bsseq_dmrs <- readRDS("../output/bsseq_dmrs.rds")
bsseq_dmrs$rank <- 1:nrow(bsseq_dmrs)
bsseq_dmrs_filt <- bsseq_dmrs %>% filter(abs(areaStat) > 100)
```
# gene source

```{r}
eh <- ExperimentHub()
eh_genes <- eh[["EH3132"]]
eh_genes <- eh_genes %>% data.frame() %>% mutate(num = 1:n())
biomart_genes <- readRDS("../output/genes_biomart.rds")
biomart_genes <- biomart_genes %>% data.frame()
```


# overlap genes and dmrs

# Distribution of gene widths
```{r}
eh_genes %>% 
  ggplot(aes(x = log2(width))) +
  geom_density()
```

# gene length and number of CpGs

eh_genes <- eh_genes %>% mutate(bin_pos = paste0(seqnames, "-", start))
counts_per_gene <- find_overlaps(as_granges(eh_genes), as_granges(wgbs_counts)) %>% data.frame()
cg_per_bin <- counts_per_gene %>% group_by(num, bin_pos) %>% summarise(n_region = n())
cg_per_bin <- cg_per_bin %>% ungroup() %>% data.frame()

eh_genes <- eh_genes %>% mutate(num_cg = cg_per_bin[match(.$num, cg_per_bin$num), "n_region"])
eh_genes[is.na(eh_genes)] <- 0



eh_genes %>%
  ggplot(aes(x = num_cg, y = width)) +
  geom_point(alpha = 0.4)

cor.test(eh_genes$num_cg, eh_genes$width)

```{r}
biomart_genes <- biomart_genes %>% mutate(width = end - start + 1)
biomart_genes  %>%
  ggplot(aes(x = num_cg, y = width)) +
  geom_point(alpha = 0.4)
```

```{r}
cor.test(biomart_genes$num_cg, biomart_genes$width)
```


## Potential biases


```{r}
source("../code/function_binBiasPlot.R")
test_anno <- annoGeneDmr(wgbs_counts, bsseq_dmrs, "biomart", biomart_genes)
test_anno_filt <- annoGeneDmr(wgbs_counts, bsseq_dmrs_filt, "biomart", biomart_genes)

overlap_gene_dmr <- find_overlaps(as_granges(bsseq_dmrs), eh_genes) %>% data.frame()
```

# Length of gene bias
```{r}
source("../code/function_binBiasPlot.R")
plotBiasGeneLength(biomart_genes, test_anno_filt, bin_size = 100)
plotBiasGeneLength(biomart_genes, test_anno_filt, bin_size = 200)

```

# Number of CpGs in Genes
```{r gene-bias}
source("../code/function_binBiasPlot.R")
plotBiasGrouped(biomart_genes, test_anno)
plotBiasGrouped(biomart_genes, test_anno_filt)

plotBiasGrouped(biomart_genes, test_anno, log2_scale = TRUE)
plotBiasGrouped(biomart_genes, test_anno_filt, log2_scale = TRUE)
```


## Analyses
```{r}
find_overlaps(as_granges(bsseq_dmrs), eh_genes) %>% data.frame() %>% group_by(bsseq_rank) %>% summarise(n_genes =n())
find_overlaps(as_granges(bsseq_dmrs), eh_genes) %>% data.frame() %>% group_by(bsseq_rank) %>% summarise(n_genes =n()) %>%
  ungroup() %>% group_by(n_genes) %>% summarise(n())

find_overlaps(as_granges(bsseq_dmrs), eh_genes) %>% data.frame() %>% group_by(symbol) %>% summarise(n_dmr =n())
find_overlaps(as_granges(bsseq_dmrs), eh_genes) %>% data.frame() %>% group_by(symbol) %>% summarise(n_dmr =n()) %>%
  ungroup() %>% group_by(n_dmr) %>% summarise(n())
```


```{r}
overlap_gene_dmr <- find_overlaps(as_granges(bsseq_dmrs), eh_genes) %>% data.frame()
```

```{r}
eh_genes %>% data.frame() %>% mutate(dmr = overlap_gene_dmr[match(.$gene, overlap_gene_dmr$gene), "gene"], 
                                     dmr = ifelse(!is.na(dmr), TRUE, FALSE)) %>%
  group_by(width, dmr) %>% summarise(n= n()) %>% group_by(width) %>% mutate(sum = sum(n), prop = n/sum, prop = ifelse(dmr == FALSE, 0, prop), num = n()) %>%
  mutate(bseq_dmr = ifelse(num == 1, TRUE, dmr)) %>% filter(dmr == TRUE) %>%
  ggplot(aes(x = width, y = prop)) +
  geom_point(alpha = 0.3)
```

```{r}
eh_genes %>% data.frame() %>% mutate(dmr = overlap_gene_dmr[match(.$gene, overlap_gene_dmr$gene), "gene"], 
                                     dmr = ifelse(!is.na(dmr), TRUE, FALSE)) %>%
  group_by(width, dmr) %>% summarise(n= n()) %>% group_by(width) %>% mutate(sum = sum(n), prop = n/sum, prop = ifelse(dmr == FALSE, 0, prop), num = n()) %>%
  mutate(bseq_dmr = ifelse(num == 1, TRUE, dmr)) %>% filter(dmr == TRUE) %>%
  ggplot(aes(x = log2(width), y = prop)) +
  geom_point(alpha = 0.3)
```

```{r}
eh_genes <- eh_genes %>% data.frame() %>% mutate(num = 1:n())
eh_genes <- eh_genes %>% mutate(bin_pos = paste0(seqnames, "-", start))
counts_per_gene <- find_overlaps(as_granges(eh_genes), as_granges(wgbs_counts)) %>% data.frame()
#eh_genes <- eh_genes %>% mutate(bin_pos = paste0(seqnames, "-", start))
cg_per_bin <- counts_per_gene %>% group_by(num, bin_pos) %>% summarise(n_region = n())
cg_per_bin <- cg_per_bin %>% ungroup() %>% data.frame()

eh_genes <- eh_genes %>% mutate(num_cg = cg_per_bin[match(.$num, cg_per_bin$num), "n_region"])
eh_genes[is.na(eh_genes)] <- 0

```


```{r}
eh_genes %>% data.frame() %>% mutate(has_dmr = overlap_gene_dmr[match(.$gene, overlap_gene_dmr$gene), "gene"], 
                                     has_dmr = ifelse(!is.na(has_dmr), TRUE, FALSE)) %>% group_by(num_cg) %>% 
  group_by(num_cg, has_dmr) %>% summarise(num_per_dmr = n()) %>% group_by(num_cg) %>% mutate(prop_dmr = num_per_dmr/sum(num_per_dmr)) %>% filter(has_dmr == TRUE)
eh_genes %>% data.frame() %>% mutate(has_dmr = overlap_gene_dmr[match(.$gene, overlap_gene_dmr$gene), "gene"], 
                                     has_dmr = ifelse(!is.na(has_dmr), TRUE, FALSE)) %>% group_by(num_cg) %>% 
  group_by(num_cg, has_dmr) %>% summarise(num_per_dmr = n()) %>% group_by(num_cg) %>% mutate(prop_dmr = num_per_dmr/sum(num_per_dmr)) %>% filter(has_dmr == TRUE) %>%
  ggplot(aes(x = num_cg, y = prop_dmr)) +
  geom_point(alpha = 0.4)
```

```{r}
eh_genes %>%
  ggplot(aes(x = num_cg, y = width)) +
  geom_point(alpha = 0.4)
```

# how many genes actually have a dmr?
```{r}
eh_genes %>% data.frame() %>% mutate(has_dmr = overlap_gene_dmr[match(.$gene, overlap_gene_dmr$gene), "gene"], 
                                     has_dmr = ifelse(!is.na(has_dmr), TRUE, FALSE)) %>%
  group_by(has_dmr) %>% summarise(n=n()) %>% ungroup() %>% mutate(prop = n/sum(n))
```

```{r}
eh_genes <- eh_genes %>% data.frame() %>% mutate(has_dmr = overlap_gene_dmr[match(.$gene, overlap_gene_dmr$gene), "gene"], 
                                     has_dmr = ifelse(!is.na(has_dmr), TRUE, FALSE))
```

# what prop of dmrs overlap to gene?

```{r}
bsseq_dmrs %>% mutate(overlap_gene = overlap_gene_dmr[match(.$bsseq_rank, overlap_gene_dmr$bsseq_rank), "gene"],
                      has_gene = ifelse(is.na(overlap_gene), FALSE, TRUE)) %>% group_by(has_gene) %>% summarise(n=n()) %>%
  ungroup() %>% mutate(prop = n/sum(n))

```


- do for width of gene??



# more detailed gene features
```{r}
library(annotatr)
```

```{r}
annots_gene <- build_annotations(genome = 'hg19', annotations = c('hg19_basicgenes', 'hg19_genes_intergenic',
    'hg19_genes_intronexonboundaries'))
```

```{r}
annots_gene <- annots_gene %>% data.frame() %>% filter(seqnames %in% unique(bsseq_res$seqnames)) %>% as_granges()
```

```{r}
#anno_cpg_by_gene <- annotate_regions(as_granges(bsseq_res), annots_gene) %>% data.frame()
```

```{r}
bseq_dmrs_genes <- annotate_regions(as_granges(bsseq_dmrs), annots_gene) %>% data.frame()
```


```{r}
bseq_dmrs_genes %>% group_by(annot.type) %>% summarise(n=n()) %>% ungroup() %>% mutate(prop = n/sum(n))
```
```{r}
bseq_dmrs_genes %>% distinct(rank, annot.type) %>% group_by(annot.type) %>% summarise(n_dmr = n(), prop_total_dmr = n()/69858)
```



# cpg annotations - ISLANDS
```{r}
annotations %>% head()
```
```{r}
anno_islands <- annotations %>% data.frame() %>% filter(type == "hg19_cpg_islands") %>% mutate(island_num = 1:n())
island_cg <- find_overlaps(as_granges(anno_islands), as_granges(wgbs_counts)) %>% data.frame() %>% group_by(island_num) %>% summarise(num_cg = n()) %>% data.frame()
```
```{r}
head(anno_islands)
```

```{r}
find_overlaps(as_granges(eh_genes), as_granges(anno_islands)) %>% data.frame() %>% group_by(gene, num_cg.x) %>% summarise(n_islands = n()) %>%
  ggplot(aes(x = num_cg.x, y = n_islands)) +
  geom_point(alpha = 0.4)
```
```{r}
find_overlaps(as_granges(eh_genes), as_granges(anno_islands)) %>% data.frame() %>%
  group_by(width) %>% summarise(n_islands = n()) %>%
  ggplot(aes(x = log2(width), y = n_islands)) +
  geom_point(alpha = 0.4)
```


```{r}
anno_islands <- anno_islands %>% mutate(num_cg = island_cg[match(.$island_num, island_cg$island_num), "num_cg"])
anno_islands %>% group_by(num_cg) %>% summarise(n= n()) %>%
  ggplot(aes(x = num_cg, y = n)) +
  geom_point()
```

```{r}
bseq_dmrs_cpg_anno <- annotate_regions(as_granges(bsseq_dmrs), annotations) %>% data.frame()
```
```{r}
bseq_dmrs_cpg_isl <- filter(bseq_dmrs_cpg_anno, annot.type == "hg19_cpg_islands")
```

```{r}
anno_islands %>% mutate(has_dmr = ifelse(id %in% bseq_dmrs_cpg_isl$annot.id, TRUE, FALSE)) %>% group_by(num_cg) %>%
      mutate(num_bins_same_cg = n()) %>%
    group_by(num_cg, has_dmr) %>%
    mutate(num_per_dmr = n(), prop = n()/num_bins_same_cg) %>%
    distinct(num_cg, num_bins_same_cg, num_per_dmr, prop, has_dmr) %>%
    group_by(num_cg) %>%
    .[order(.$num_cg),] %>%
    mutate(prop = ifelse(has_dmr == FALSE, 0, prop)) %>%
    mutate(num = n()) %>%
    mutate(has_dmr = ifelse(num == 1, TRUE, has_dmr)) %>% filter(has_dmr == TRUE) %>%
    ggplot(aes(x = num_cg, y = prop)) +
      geom_point(alpha = 0.4) +
      geom_smooth()
```


- prop of dmrs that have island
```{r}
bsseq_dmrs %>% mutate(has_island = ifelse(rank %in% bseq_dmrs_cpg_isl$rank, TRUE, FALSE)) %>% group_by(has_island) %>% summarise(n=n()) %>% ungroup() %>% mutate(prop = n/sum(n))
```

#################
# repeat for biomart (has a tss)



```{r}
anno_biomart_dmr %>%
  ggplot(aes(x = min_dist)) +
  geom_density()
```


# cpg density per gene
```{r}
biomart_genes %>% mutate(width = end - start + 1, density = num_cg/width)
biomart_genes %>% mutate(width = end - start + 1, density = num_cg/width) %>% group_by(density) %>% summarise(n_match = n())
biomart_genes %>% mutate(width = end - start + 1, density = num_cg/width) %>% 
  group_by(density) %>% summarise(n_match = n()) %>%
  ggplot(aes(x = density, y = n_match)) +
  geom_point(alpha = 0.3)
```
- what was i supposed to do with density??
- is it about if more dense genes have dmr?
```{r}
biomart_genes <- biomart_genes %>% mutate(width = end - start + 1, density = num_cg/width)
biomart_genes <- biomart_genes %>% mutate(overlap_dmr = ifelse(ensembl_gene_id %in% filter(anno_biomart_dmr, abs(min_dist) == 0)$ensembl_gene_id, TRUE, FALSE))
```

```{r}
biomart_genes %>%
  ggplot(aes(x = overlap_dmr, y = density, colour = overlap_dmr)) +
  geom_boxplot()
biomart_genes %>%
  ggplot(aes(x = overlap_dmr, y = log2(density + 0.1), colour = overlap_dmr)) +
  geom_boxplot()
```
```{r}
t.test(filter(biomart_genes, overlap_dmr == TRUE)$density, filter(biomart_genes, overlap_dmr == FALSE)$density)
t.test(filter(biomart_genes, overlap_dmr == TRUE)$density, filter(biomart_genes, overlap_dmr == FALSE)$density, alternative = "less")
t.test(filter(biomart_genes, overlap_dmr == TRUE)$density, filter(biomart_genes, overlap_dmr == FALSE)$density, alternative = "greater")
```
- nope p vals are not signif
- what else could do with density??
- oh was this connected to the promoter stuff where we want to know if they cluster more at start of gene than end?
- i think density comes into that
- i think it was to do with promoter and 1st exon?


##
```{r}
biomart_genes %>%
  ggplot(aes(x = log2(num_cg + 1))) +
  geom_density()
```




