---
title: "03a_developGeneSetTest"
author: "Caitlin Page"
date: "2024-11-15"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: inline
---
```{r}
library(dplyr)
library(plyranges)
library(missMethyl)
library(GO.db)
library(ggplot2)
```

## Introduction

```{r}

```

# i want to compare the gene sets
- figure out if that's what's going wrong
(and maybe paste all my code in here lol)
or just rename the other file to this


## get the gene set from goseq
goseq::goseq(pwf, "hg19", "ensGene")
getgo(rownames(pwf),genome,id,fetch.cats=test.cats)
```{r}
rownames(pwf) %>% length()
test_bias$ensembl_gene_id %>% length()
```

```{r}
go_goseq <- getgo(test_bias$ensembl_gene_id, "hg19", "ensGene")
names(go_goseq) <- test_bias$ensembl_gene_id
go_goseq <- reversemapping(go_goseq)
go_goseq <- reversemapping(go_goseq)
goseq_go <- reversemapping(go_goseq)
go_goseq
```
```{r}
names(goseq_go)[1:5]
names(go_goseq)[1:5]
```

```{r}
length(unique(names(goseq_go))) # goseq 22258 
length(unique(names(collection))) # mm 22124
```

```{r}
ggVennDiagram(list(mm = unique(names(collection)), goseq = unique(names(goseq_go))))
```


```{r}
names(goseq_go[test_bias$ensembl_gene_id]) %>% unique() %>% length()
length(test_bias$ensembl_gene_id) - length(go_goseq)
```
```{r}
test_bias$ensembl_gene_id %in% names(go_goseq) %>% summary()
go_goseq[filter(test_bias, ensembl_gene_id %in% names(go_goseq))$ensembl_gene_id] %>% length()
```
```{r}
go_goseq_filt <- go_goseq[filter(test_bias, ensembl_gene_id %in% names(go_goseq))$ensembl_gene_id]
goseq_go_filt <- reversemapping(go_goseq_filt)
```
```{r}
length(goseq_go_filt)
names(goseq_go_filt)[1:5]
```

- so I filtered out the genes but it didn't change the categories
```{r}
lapply(go_goseq[1:5], length)
lapply(goseq_go[1:5], length)
```
```{r}
lapply(goseq_go, length) %>% unlist() %>% data.frame() %>% setNames("num_genes") #22258, same with filt
lapply(goseq_go_filt, length) %>% unlist() %>% data.frame() %>% setNames("num_genes") %>% filter(num_genes > 0)
```
there's no diff between these
```{r}
ggVennDiagram(list(mm = unique(names(collection)), goseq = unique(names(goseq_go_filt))))
```

- ok so this can't be the cause of getting more res in mm - also they are fairly sim

- so what is happening?
-running with the other inputs did not fix it
##############
# functions
```{r}
# copied from mm
.getGO <- function(){
  if(!requireNamespace("org.Hs.eg.db", quietly = TRUE))
    stop("org.Hs.eg.db package required but not installed.")
  egGO2ALLEGS <- utils::getFromNamespace("org.Hs.egGO2ALLEGS", "org.Hs.eg.db")
  GeneID.PathID <- AnnotationDbi::toTable(egGO2ALLEGS)[,c("gene_id", "go_id", "Ontology")]
  d <- !duplicated(GeneID.PathID[, c("gene_id", "go_id")])
  GeneID.PathID <- GeneID.PathID[d, ]
  GOID.TERM <- suppressMessages(AnnotationDbi::select(GO.db::GO.db, 
                                                      keys=unique(GeneID.PathID$go_id), 
                                                      columns=c("GOID","ONTOLOGY","TERM"), 
                                                      keytype="GOID"))
  go <- tapply(GeneID.PathID$gene_id, GeneID.PathID$go_id, list)
    
  list(idList=go, idTable=GOID.TERM)
}
go <- .getGO()

.estimatePWF <- function(D,bias)
  # An alternative to goseq function nullp, which is transformation invariant
  # Belinda Phipson and Gordon Smyth
  # 6 March 2015
{
  prior.prob <- bias
  o <- order(bias)
  prior.prob[o] <- limma::tricubeMovingAverage(D[o],span=0.5)
  prior.prob
}
test_pwf <- .estimatePWF(D=test.de,bias=as.vector(freq_genes))

.plotBias <- function(D,bias)
  # Plotting function to show gene level CpG density bias
  # Belinda Phipson
  # 5 March 2015
{
  o <- order(bias)
  splitf <- rep(1:100,each=200)[1:length(bias)]
  avgbias <- tapply(bias[o],factor(splitf),mean)
  sumDM <- tapply(D[o],factor(splitf),sum)
  propDM <- sumDM/table(splitf)
  graphics::par(mar=c(5,5,2,2))
  graphics::plot(avgbias,as.vector(propDM),
                 xlab="Number of CpGs per gene (binned)",
                 ylab="Proportion Differential Methylation",cex.lab=1.5,
                 cex.axis=1.2)
  graphics::lines(stats::lowess(avgbias,propDM),col=4,lwd=2)
}
.plotBias(D=test.de,bias=as.vector(freq_genes))
```
```{r}
library(BiasedUrn)
```
```{r}
biomart_genes %>% mutate(width = end - start + 1, has_dmr = ifelse(ensembl_gene_id %in% filter(test_anno, abs(min_dist) == 0)$ensembl_gene_id, TRUE, FALSE)) %>%
  group_by(num_cg) %>%
    mutate(num_bins_same_cg = n()) %>% 
  .[order(.$num_cg),] %>%
  ungroup() %>%
  mutate(bin_group = rep(1:ceiling(nrow(biomart_genes)/100), each = 100)[1:nrow(biomart_genes)]) %>% 
  group_by(bin_group) %>%
  mutate(num_cg_bin_group = mean(num_cg)) %>%
  group_by(num_cg_bin_group) %>%
    mutate(num_bins_group_same_cg = dplyr::n()) %>%
    group_by(num_cg_bin_group, has_dmr) %>%
    mutate(num_per_dmr = dplyr::n(), prop = dplyr::n()/num_bins_group_same_cg) %>%
    distinct(num_cg_bin_group, num_bins_group_same_cg, num_per_dmr, prop, has_dmr) %>%
    group_by(num_cg_bin_group) %>%
    .[order(.$num_cg_bin_group),] %>%
    mutate(prop = ifelse(has_dmr == FALSE, 0, prop)) %>%
    mutate(num = dplyr::n()) %>%
    mutate(has_dmr = ifelse(num == 1, TRUE, has_dmr)) %>% filter(has_dmr == TRUE)
```
##############
# functions
```{r}
biomart_genes <- readRDS("../output/genes_biomart.rds")
bsseq_dmrs <- readRDS("../output/bsseq_dmrs.rds")
bsseq_res <- readRDS("../output/bsseq_res.rds")
```
```{r}
test_anno <- annoGeneDmr(bsseq_res, bsseq_dmrs, gene_source = "biomaRt", biomart_genes)
test_anno
```

##########


- weird i don't think the pwf is even used at all
- yes it is test_pwf is part of the testing
therefore the thing that is taking forever is not going to run


# functions again
I have
```{r}
sourceGenes() # takes 'counts' - for chrom positions and num_cg
annoGeneDmr() # as it sounds

.getGO() # stolen from mm
.estimatePWF() # stolen from mm
```
then my run_mm a and b

If I want to redo how talked in meeting - then basically all those things would be wrapped
and the testing
so the individual functions can stay, but we would have 1 thing that would just do it all
```{r}
run_gst_seq <- function(dmrs, tested_cpgs, 
                        gene_source = c("biomaRt", "ExperimentHub"), gene_feature=c("gene", "promoter"),
                        correct_bias=TRUE, plot_bias=TRUE, ...) {
  gene_source <- match.arg(tolower(gene_source), c("biomart", "experimenthub"))
  gene_feature <- match.arg(tolower(gene_feature), c("gene", "promoter"))
  genes <- sourceGenes(counts = tested_cpgs, gene_source = gene_source, gene_feature = gene_feature)
  if(nrow(genes) == 0) {
    stop("sourceGenes failed")
  }
  if(nrow(dmrs) > 30000) {
    stop("Please appropriately filter DMRs to most significant") #idk
  }
  anno <- annoGeneDmr(counts=tested_cpgs, dmrs = dmrs, gene_source = gene_source, genes = genes)
  if(nrow(anno) == 0) {
    stop("anno failed")
  }
out <- run_miss_methyl_a(genes, anno)
  results <- run_miss_methyl_b(out, correct_bias=correct_bias, plot_bias=plot_bias, ...)

  list(res = results, genes = genes, anno = anno, out = out)
}
test_fn_res <- run_gst_seq(bsseq_dmrs[1:1000,], bsseq_res, "biomaRt", "gene")
```

```{r}
test_out_again <- run_miss_methyl_a(test_fn_res$genes, test_fn_res$anno)
```

```{r}
test_fn_res
```

```{r}
system.time(sourceGenes(bsseq_res, "biomart", "gene"))
system.time(annoGeneDmr(bsseq_res, bsseq_dmrs[1:1000,], "biomart", test_fn_res$genes))
system.time(run_miss_methyl_a(test_fn_res$genes, test_fn_res$anno))
system.time(run_miss_methyl_b(test_out_again, correct_bias = TRUE))
system.time(run_miss_methyl_b(test_out_again, correct_bias = FALSE))
```


```{r}
run_gst_seq(bsseq_dmrs[1:1000], bsseq_res, "biomaRt", "gene")
```

```{r}
bsseq_res <- readRDS("../output/bsseq_res.rds")
```

```{r}
?toupper
?match.arg
```
```{r}
mm_hack_res[mm_hack_res$ONTOLOGY %in% c("MF"),]
```
```{r}
biomart_genes %>% mutate(a = NA) %>% .[!.$a %in% NA,]
biomart_genes[biomart_genes$num_cg > 0,] %>% .[order(.$num_cg),]
biomart_genes <- readRDS("../output/genes_biomart.rds")
```

```{r}
run_miss_methyl_a <- function(biomart_genes, test_anno) { # add something about extending the bias to include upstream of gene maybe??
  entrez_genes <- AnnotationDbi::select(org.Hs.eg.db, 
                                keys=AnnotationDbi::keys(org.Hs.eg.db), 
                                columns=c("ENTREZID","SYMBOL"), 
                                keytype="ENTREZID")
  biomart_genes$entrez_id <- entrez_genes[match(biomart_genes$gene_name, entrez_genes$SYMBOL), "ENTREZID"]
  biomart_genes_filt <- biomart_genes[!biomart_genes$entrez_id %in% NA,]
  biomart_genes_filt <- biomart_genes_filt %>% group_by(gene_name) %>% filter(num_cg == max(num_cg)) %>% ungroup() #this line will be slow
  
  test_anno$entrez_id <- entrez_genes[match(test_anno$gene_name, entrez_genes$SYMBOL), "ENTREZID"]
  test_anno_filt <- test_anno[!test_anno$entrez_id %in% NA,] 
  test_anno_filt <- test_anno_filt[test_anno_filt$tss_rel_peak %in% "Overlap",]
  biomart_genes_filt$bias <- biomart_genes_filt$num_cg
  
  
    
  test_uni_freq <- biomart_genes_filt %>% .[order(.$entrez_id),] %>% group_by(entrez_id) %>% distinct(entrez_id, bias) %>% ungroup() %>% data.frame() # probs slow
  test_uni_freq <- test_uni_freq[test_uni_freq$bias > 0,]
  test_uni_freq

  freq_make_table <- rep(test_uni_freq$entrez_id, test_uni_freq$bias) #this has replaced the loop and sped it up

  freq_make_table <- table(freq_make_table)
  biomart_genes_filt <- biomart_genes_filt[biomart_genes_filt$bias > 0,] 
  test_anno_filt <- test_anno_filt[test_anno_filt$ensembl_gene_id %in% biomart_genes_filt$ensembl_gene_id,]
  biomart_genes_filt$is_de <- ifelse(biomart_genes_filt$entrez_id %in% test_anno_filt$entrez_id, 1, 0)
  biomart_genes_filt$is_de <- as.integer(biomart_genes_filt$is_de)
  out <- list(
  #fract.counts = biomart_genes_filt %>%
   # mutate(sigid = entrez_id) %>% distinct(sigid) %>% .[order(.$sigid),] %>% data.frame() %>% 
   # mutate(frac = as.double(1)),
  sig.eg = test_anno_filt %>% distinct(entrez_id) %>% .[order(.$entrez_id),],
  universe = biomart_genes_filt %>% distinct(entrez_id) %>% 
    .[order(.$entrez_id),] %>% .$entrez_id,
  freq = freq_make_table,
  equiv = as.array(freq_make_table), # column names wrong but maybe not too bad?
  de = biomart_genes_filt[order(biomart_genes_filt$entrez_id),] %>% .$is_de
)
  out
}
system.time(run_miss_methyl_a(biomart_genes, test_anno))
#test_out <- run_miss_methyl_a(biomart_genes, test_anno)# ; beep(2)
```
- think screw up was here - missed a filter so rerunning






```{r}
# filter the dmr
test_anno %>% filter(abs(areaStat) > 100) #it's like 17k instead of 70k
test_out_filt <- run_miss_methyl_a(biomart_genes, filter(test_anno, abs(areaStat) > 100))
```
```{r}
library(org.Hs.eg.db)
```

```{r}
test_out_num <- run_miss_methyl_a(biomart_genes, test_anno, "num_cg")
```
```{r}
test_out <- readRDS("../output/mm_fn_out_num.rds")
```


- something goes wrong with the frac.counts - but do I need that downstream?
- no I don't so it can be culled
- ok it ran - it took like 40 minutes which is utterly ridiculous - maybe it's just because of the size of the data? - but still, I feel like it was faster when I was just running the bits of code individually to get it running
- now for part b of it
```{r}
go <- .getGO()
```

```{r}
run_miss_methyl_b <- function(out, correct_bias=TRUE, plot_bias=FALSE, ...) {
  sorted.eg.sig <- out$sig.eg #entrez ids for sig cpgs - genes that have sig cpgs
  eg.universe <- out$universe # entrez ids all cpgs(?) - but it's just all genes
  freq_genes <- out$freq
  test.de <- out$de #res for each gene
  #frac <- out$fract.counts
  equiv <- out$equiv
  ## setting up stuff
  # Check collection is a list with character vectors
  go <- .getGO()
  collection <- go$idList  
  #only if it's not a list
  #collection <- list(collection=go$idList)
  collection <- lapply(collection, as.character)
  # Make sure gene set collections don't have any NAs
  collection <- lapply(collection, function(x) x[!is.na(x)])
  # Remove genes that are NOT in the universe from collections
  collection <- lapply(collection, function(x) x[x %in% eg.universe])
  # Remove collections with no genes left after universe filter
  inUniv <- sapply(collection, function(x) length(x) > 0)
  collection <- collection[inUniv]
  
  
  results <- matrix(NA,ncol=4,nrow=length(collection))
  colnames(results) <- c("N", "DE", "P.DE","FDR")
  rownames(results) <- names(collection)
  results[,"N"] <- unlist(lapply(collection,length))
  SigGenesInSet <- rep(NA,length(collection)) # this one is optional
  
  results[,"DE"] <- unlist(lapply(collection, function(x) 
          sum((sorted.eg.sig %in% x))))
  Nuniverse <- length(eg.universe)
  m <- length(sorted.eg.sig)
  
  if(correct_bias == TRUE) {
  test_pwf <- .estimatePWF(D=test.de,bias=as.vector(freq_genes))
  #  if(plot_bias == TRUE) {
   #   plotBiasGrouped(genes = genes, anno_genes = , ...)
  #  }
  # the testing
  results[, c("P.DE")] <- 
  t(sapply(1:length(collection), function(i) {
    InSet <- eg.universe %in% collection[[i]]
    pw.red <- sum(test_pwf[InSet], na.rm = TRUE) / results[i, "N"] # so i don't know why I get na's - or do i?
    pw.white <- sum(test_pwf[!InSet], na.rm = TRUE) / (Nuniverse - results[i, "N"]) #anyway the na removal is not in original code
    odds <- pw.red / pw.white
    p_val <- BiasedUrn::pWNCHypergeo(results[i,"DE"],
                                                   results[i,"N"],
                                                   Nuniverse-results[i,"N"],
                                                   m,odds,lower.tail=FALSE) + 
          BiasedUrn::dWNCHypergeo(results[i,"DE"],
                                  results[i,"N"],
                                  Nuniverse-results[i,"N"],
                                  m,odds)
    #the optional thing
      # Get gene symbols of significant genes
    SigGenesEntrezID <- sorted.eg.sig[sorted.eg.sig %in% collection[[i]]]
    SigGenesSymbol <- suppressMessages(AnnotationDbi::select(org.Hs.eg.db, 
                                             keys = SigGenesEntrezID,
                                             columns = "SYMBOL"))
    SigGenesInSet[i] <- paste(SigGenesSymbol$SYMBOL,collapse=",")
    if(p_val==0){
        message("Pvalue of exactly zero detected. Performing hypergeometric 
                test for gene set ", rownames(results)[i])
        p_val <- stats::phyper(q=results[i,"DE"]-0.5,m=m,
                                           n=Nuniverse-m,k=results[i,"N"],
                                           lower.tail=FALSE)
    }
    c(p_val)  # return the values as a vector
    })) #chatgpt translated loop to apply
  } else {
    results[, c("P.DE")] <- 
    t(sapply(1:length(collection), function(i) {
      p_val <- stats::phyper(q=results[i,"DE"]-0.5,m=m,
                                           n=Nuniverse-m,k=results[i,"N"],
                                           lower.tail=FALSE)
      #the optional thing
      # Get gene symbols of significant genes
    SigGenesEntrezID <- sorted.eg.sig[sorted.eg.sig %in% collection[[i]]]
    SigGenesSymbol <- suppressMessages(AnnotationDbi::select(org.Hs.eg.db, 
                                             keys = SigGenesEntrezID,
                                             columns = "SYMBOL"))
    SigGenesInSet[i] <- paste(SigGenesSymbol$SYMBOL,collapse=",")
      c(p_val)
    }))
  }
  
  
  results <- results %>% data.frame()
  results[,"FDR"] <- stats::p.adjust(results[,"P.DE"],method="BH")
  results[,"DE"] <- floor(results[,"DE"])
  results <- data.frame(results, SigGenesInSet)
  
  results <- merge(go$idTable,results,by.x="GOID",by.y="row.names")
  rownames(results) <- results$GOID
  results <- results %>% .[order(.$FDR),]
  results
  
}

test_b_num_unfilt <- run_miss_methyl_b(test_out)
test_b_num_null_unfilt <- run_miss_methyl_b(test_out, correct_bias = FALSE)
```
```{r}
just_test <- function(out) {
  sorted.eg.sig <- out$sig.eg #entrez ids for sig cpgs - genes that have sig cpgs
  eg.universe <- out$universe # entrez ids all cpgs(?) - but it's just all genes
  freq_genes <- out$freq
  test.de <- out$de #res for each gene
  #frac <- out$fract.counts
  equiv <- out$equiv
  ## setting up stuff
  # Check collection is a list with character vectors
  go <- .getGO()
  collection <- go$idList  
  #only if it's not a list
  #collection <- list(collection=go$idList)
  collection <- lapply(collection, as.character)
  # Make sure gene set collections don't have any NAs
  collection <- lapply(collection, function(x) x[!is.na(x)])
  # Remove genes that are NOT in the universe from collections
  collection <- lapply(collection, function(x) x[x %in% eg.universe])
  # Remove collections with no genes left after universe filter
  inUniv <- sapply(collection, function(x) length(x) > 0)
  collection <- collection[inUniv]
  
  
  results <- matrix(NA,ncol=4,nrow=length(collection))
  colnames(results) <- c("N", "DE", "P.DE","FDR")
  rownames(results) <- names(collection)
  results[,"N"] <- unlist(lapply(collection,length))
  SigGenesInSet <- rep(NA,length(collection)) # this one is optional
  
  results[,"DE"] <- unlist(lapply(collection, function(x) 
          sum((sorted.eg.sig %in% x))))
  Nuniverse <- length(eg.universe)
  m <- length(sorted.eg.sig)
  
  test_pwf <- .estimatePWF(D=test.de,bias=as.vector(freq_genes))
  
  results[, c("P.DE")] <- 
  t(sapply(1:length(collection), function(i) {
    InSet <- eg.universe %in% collection[[i]]
    pw.red <- sum(test_pwf[InSet], na.rm = TRUE) / results[i, "N"] # so i don't know why I get na's - or do i?
    pw.white <- sum(test_pwf[!InSet], na.rm = TRUE) / (Nuniverse - results[i, "N"]) #anyway the na removal is not in original code
    odds <- pw.red / pw.white
    p_val <- BiasedUrn::pWNCHypergeo(results[i,"DE"],
                                                   results[i,"N"],
                                                   Nuniverse-results[i,"N"],
                                                   m,odds,lower.tail=FALSE) + 
          BiasedUrn::dWNCHypergeo(results[i,"DE"],
                                  results[i,"N"],
                                  Nuniverse-results[i,"N"],
                                  m,odds)
    #the optional thing
      # Get gene symbols of significant genes
    SigGenesEntrezID <- sorted.eg.sig[sorted.eg.sig %in% collection[[i]]]
    SigGenesSymbol <- suppressMessages(AnnotationDbi::select(org.Hs.eg.db, 
                                             keys = SigGenesEntrezID,
                                             columns = "SYMBOL"))
    SigGenesInSet[i] <- paste(SigGenesSymbol$SYMBOL,collapse=",")
    if(p_val==0){
        message("Pvalue of exactly zero detected. Performing hypergeometric 
                test for gene set ", rownames(results)[i])
        p_val <- stats::phyper(q=results[i,"DE"]-0.5,m=m,
                                           n=Nuniverse-m,k=results[i,"N"],
                                           lower.tail=FALSE)
    }
    c(p_val)  # return the values as a vector
    }))
  results
}
```

```{r}
just_res_loop <- function(out) {
  sorted.eg.sig <- out$sig.eg #entrez ids for sig cpgs - genes that have sig cpgs
  eg.universe <- out$universe # entrez ids all cpgs(?) - but it's just all genes
  freq_genes <- out$freq
  test.de <- out$de #res for each gene
  #frac <- out$fract.counts
  equiv <- out$equiv
  ## setting up stuff
  # Check collection is a list with character vectors
  go <- .getGO()
  collection <- go$idList  
  #only if it's not a list
  #collection <- list(collection=go$idList)
  collection <- lapply(collection, as.character)
  # Make sure gene set collections don't have any NAs
  collection <- lapply(collection, function(x) x[!is.na(x)])
  # Remove genes that are NOT in the universe from collections
  collection <- lapply(collection, function(x) x[x %in% eg.universe])
  # Remove collections with no genes left after universe filter
  inUniv <- sapply(collection, function(x) length(x) > 0)
  collection <- collection[inUniv]
  
  
  results <- matrix(NA,ncol=4,nrow=length(collection))
  colnames(results) <- c("N", "DE", "P.DE","FDR")
  rownames(results) <- names(collection)
  results[,"N"] <- unlist(lapply(collection,length))
  SigGenesInSet <- rep(NA,length(collection)) # this one is optional
  
  results[,"DE"] <- unlist(lapply(collection, function(x) 
          sum((sorted.eg.sig %in% x))))
  Nuniverse <- length(eg.universe)
  m <- length(sorted.eg.sig)
  

  pwf <- .estimatePWF(D=test.de,bias=as.vector(freq_genes))
  

  for(i in 1:length(collection)){
      InSet <- eg.universe %in% collection[[i]]
      pw.red <- sum(pwf[InSet])/results[i,"N"]
      pw.white <- sum(pwf[!InSet])/(Nuniverse-results[i,"N"])
      odds <- pw.red/pw.white
      results[i,"P.DE"] <- BiasedUrn::pWNCHypergeo(results[i,"DE"],
                                                   results[i,"N"],
                                                   Nuniverse-results[i,"N"],
                                                   m,odds,lower.tail=FALSE) + 
          BiasedUrn::dWNCHypergeo(results[i,"DE"],
                                  results[i,"N"],
                                  Nuniverse-results[i,"N"],
                                  m,odds)
      
        # Get gene symbols of significant genes
        SigGenesEntrezID <- sorted.eg.sig[sorted.eg.sig %in% collection[[i]]]
        SigGenesSymbol <- suppressMessages(AnnotationDbi::select(org.Hs.eg.db, 
                                             keys = SigGenesEntrezID,
                                             columns = "SYMBOL"))
        SigGenesInSet[i] <- paste(SigGenesSymbol$SYMBOL,collapse=",")
      
      if(results[i,"P.DE"]==0){
        message("Pvalue of exactly zero detected. Performing hypergeometric 
                test for gene set ", rownames(results)[i])
        results[i,"P.DE"] <- stats::phyper(q=results[i,"DE"]-0.5,m=m,
                                           n=Nuniverse-m,k=results[i,"N"],
                                           lower.tail=FALSE)
      }
  }
  results
}
just_res_loop(test_out_again)
```
```{r}
just_res_loop2 <- function(out) {
  sorted.eg.sig <- out$sig.eg #entrez ids for sig cpgs - genes that have sig cpgs
  eg.universe <- out$universe # entrez ids all cpgs(?) - but it's just all genes
  freq_genes <- out$freq
  test.de <- out$de #res for each gene
  #frac <- out$fract.counts
  equiv <- out$equiv
  ## setting up stuff
  # Check collection is a list with character vectors
  go <- .getGO()
  collection <- go$idList  
  #only if it's not a list
  #collection <- list(collection=go$idList)
  collection <- lapply(collection, as.character)
  # Make sure gene set collections don't have any NAs
  collection <- lapply(collection, function(x) x[!is.na(x)])
  # Remove genes that are NOT in the universe from collections
  collection <- lapply(collection, function(x) x[x %in% eg.universe])
  # Remove collections with no genes left after universe filter
  inUniv <- sapply(collection, function(x) length(x) > 0)
  collection <- collection[inUniv]

results <- matrix(NA,ncol=4,nrow=length(collection))
  colnames(results) <- c("N", "DE", "P.DE","FDR")
  rownames(results) <- names(collection)
  results[,"N"] <- unlist(lapply(collection,length))
  SigGenesInSet <- rep(NA,length(collection)) # this one is optional
  
  results[,"DE"] <- unlist(lapply(collection, function(x) 
          sum((sorted.eg.sig %in% x))))
  Nuniverse <- length(eg.universe)
  m <- length(sorted.eg.sig)
  

  pwf <- .estimatePWF(D=test.de,bias=as.vector(freq_genes))
  

  for(i in 1:length(collection)){
      InSet <- eg.universe %in% collection[[i]]
      pw.red <- sum(pwf[InSet])/results[i,"N"]
      pw.white <- sum(pwf[!InSet])/(Nuniverse-results[i,"N"])
      odds <- pw.red/pw.white
      results[i,"P.DE"] <- BiasedUrn::pWNCHypergeo(results[i,"DE"],
                                                   results[i,"N"],
                                                   Nuniverse-results[i,"N"],
                                                   m,odds,lower.tail=FALSE) + 
          BiasedUrn::dWNCHypergeo(results[i,"DE"],
                                  results[i,"N"],
                                  Nuniverse-results[i,"N"],
                                  m,odds)
      
        # Get gene symbols of significant genes
        SigGenesEntrezID <- sorted.eg.sig[sorted.eg.sig %in% collection[[i]]]
        SigGenesSymbol <- suppressMessages(AnnotationDbi::select(org.Hs.eg.db, 
                                             keys = SigGenesEntrezID,
                                             columns = "SYMBOL"))
        SigGenesInSet[i] <- paste(SigGenesSymbol$SYMBOL,collapse=",")
      
      if(results[i,"P.DE"]==0){
        message("Pvalue of exactly zero detected. Performing hypergeometric 
                test for gene set ", rownames(results)[i])
        results[i,"P.DE"] <- stats::phyper(q=results[i,"DE"]-0.5,m=m,
                                           n=Nuniverse-m,k=results[i,"N"],
                                           lower.tail=FALSE)
      }
  }
  results
}
```
```{r}
collection
```


```{r}
system.time(lapply(collection, function(x) test_out_again$universe %in% x))
system.time(for(i in 1:length(collection)){
      InSet <- test_out_again$universe %in% collection[[i]]})
```
I'm not convinced this difference will change much with iterations

```{r}
1+2
```

```{r}
test_b_num_dmr <- run_miss_methyl_b(test_out_filt)
test_b_num_dmr_null <- run_miss_methyl_b(test_out_filt, correct_bias = FALSE)
```

```{r}
mm_hack
sig_terms_array
```
```{r}
ggVennDiagram(list(seq=mm_hack$GOID, array=rownames(sig_terms_array)))
```

```{r}
test_b_num
test_b_num_null
```



```{r}
test_b[,"FDR"] <- stats::p.adjust(test_b[,"P.DE"],method="BH")
test_b[,"DE"] <- floor(test_b[,"DE"])
test_b <- test_b %>% filter(FDR < 0.01)
test_b %>% .[order(.$FDR),]
```
- yay I have results
- it is still too many 
- even worse - i have more than i did before
I think I know why - I think it's because the order got messed around 
wait no when I make the list that gets fixed so it can't be that
-yuck also got some with fdr of 0 - something must have gone wrong there
- my old hacked version used the num cg, instead of the binning - maybe that's the cause??
- also I think in hacked version I used the array definition for number of cpgs as well? - or for something?
- oh no I need to run a null version of this one

```{r}
goseq_res <- readRDS("../output/goseq_res.rds")
mm_hack_res <- readRDS("../output/mm_hack_res.rds")
```

```{r}
goseq_res
mm_hack_res
```
```{r}
test_b
test_b_null
```
- well this is a bigger problem
- the null has less significance
```{r}
test_b_num
test_b_num_null
```
- this version does have the corrected with less terms
- but none of the versions are saying anything about immune cells
```{r}
ggVennDiagram(list(fn = rownames(test_b), goseq = goseq_res$category))
ggVennDiagram(list(fn = rownames(test_b), hack = rownames(mm_hack_res)))
ggVennDiagram(list(fn = rownames(test_b), null = rownames(test_b_null)))
```
- ok so it does overlap with the other versions but why is it making so many more than my dumb hacked version??
- it should be doing the same stuff??

```{r}
ggVennDiagram(list(fn = rownames(test_b), fn_num = rownames(test_b_num)))
ggVennDiagram(list(fn_null = rownames(test_b_null), fn_num_null = rownames(test_b_num_null)))
ggVennDiagram(list(fn = rownames(test_b), fn_num_null = rownames(test_b_num_null)))
ggVennDiagram(list(fn_null = rownames(test_b_null), fn_num = rownames(test_b_num)))
```

```{r}
test_out <- readRDS("../output/mm_fn_out.rds")
test_out$
```


was slow though
for(i in 1:length(collection)){
      InSet <- eg.universe %in% collection[[i]]
      pw.red <- sum(test_pwf[InSet])/results[i,"N"]
      pw.white <- sum(test_pwf[!InSet])/(Nuniverse-results[i,"N"])
      odds <- pw.red/pw.white
      results[i,"P.DE"] <- BiasedUrn::pWNCHypergeo(results[i,"DE"],
                                                   results[i,"N"],
                                                   Nuniverse-results[i,"N"],
                                                   m,odds,lower.tail=FALSE) + 
          BiasedUrn::dWNCHypergeo(results[i,"DE"],
                                  results[i,"N"],
                                  Nuniverse-results[i,"N"],
                                  m,odds)
      #the optional thing
        # Get gene symbols of significant genes
        SigGenesEntrezID <- sorted.eg.sig[sorted.eg.sig %in% collection[[i]]]
        SigGenesSymbol <- suppressMessages(AnnotationDbi::select(org.Hs.eg.db, 
                                             keys = SigGenesEntrezID,
                                             columns = "SYMBOL"))
        SigGenesInSet[i] <- paste(SigGenesSymbol$SYMBOL,collapse=",")
      #end optional gene thing
      if(results[i,"P.DE"]==0){
        message("Pvalue of exactly zero detected. Performing hypergeometric 
                test for gene set ", rownames(results)[i])
        results[i,"P.DE"] <- stats::phyper(q=results[i,"DE"]-0.5,m=m,
                                           n=Nuniverse-m,k=results[i,"N"],
                                           lower.tail=FALSE)
      }
  }
  #res for a
  results <- matrix(NA,ncol=5,nrow=length(collection))
+   colnames(results) <- c("N", "DE", "pw.red","pw.white","odds")
+   rownames(results) <- names(collection)
+   results[,"N"] <- unlist(lapply(collection,length))
+   SigGenesInSet <- rep(NA,length(collection)) # this one is optional
+   
+   results[,"DE"] <- unlist(lapply(collection, function(x) 
+           sum((sorted.eg.sig %in% x))))
+   Nuniverse <- length(eg.universe)
+   m <- length(sorted.eg.sig)
+   test_pwf <- .estimatePWF(D=test.de,bias=as.vector(freq_genes))
+   # the testing
+   results[, c("pw.red", "pw.white", "odds")] <- 
+   t(sapply(1:length(collection), function(i) {
+     InSet <- eg.universe %in% collection[[i]]
+     pw.red <- sum(test_pwf[InSet], na.rm = TRUE) / results[i, "N"]
+     pw.white <- sum(test_pwf[!InSet], na.rm = TRUE) / (Nuniverse - results[i, "N"])
+     odds <- pw.red / pw.white
+     c(pw.red, pw.white, odds)  # return the values as a vector
+   })) #chatgpt translated loop to apply
+   results %>% data.frame()

translation 2 for whole section
results <- do.call(rbind, lapply(1:length(collection), function(i) {
  InSet <- eg.universe %in% collection[[i]]
  pw.red <- sum(test_pwf[InSet]) / results[i, "N"]
  pw.white <- sum(test_pwf[!InSet]) / (Nuniverse - results[i, "N"])
  odds <- pw.red / pw.white
  
  p_value <- BiasedUrn::pWNCHypergeo(results[i, "DE"],
                                      results[i, "N"],
                                      Nuniverse - results[i, "N"],
                                      m, odds, lower.tail = FALSE) +
             BiasedUrn::dWNCHypergeo(results[i, "DE"],
                                      results[i, "N"],
                                      Nuniverse - results[i, "N"],
                                      m, odds)
  
  # Optional gene part
  SigGenesEntrezID <- sorted.eg.sig[sorted.eg.sig %in% collection[[i]]]
  SigGenesSymbol <- suppressMessages(AnnotationDbi::select(org.Hs.eg.db,
                                                           keys = SigGenesEntrezID,
                                                           columns = "SYMBOL"))
  SigGenesInSet <- paste(SigGenesSymbol$SYMBOL, collapse = ",")
  
  # Check for p-value equal to zero and re-calculate if needed
  if (p_value == 0) {
    message("Pvalue of exactly zero detected. Performing hypergeometric test for gene set ", rownames(results)[i])
    p_value <- stats::phyper(q = results[i, "DE"] - 0.5, m = m,
                             n = Nuniverse - m, k = results[i, "N"],
                             lower.tail = FALSE)
  }
  
  # Return a row as a data frame
  results_row <- results[i, , drop = FALSE]
  results_row["P.DE"] <- p_value
  results_row["SigGenesInSet"] <- SigGenesInSet
  
  return(results_row)
}))


results <- results


- if b works - readd the rest of the code to it
- hey this part was pretty fast - the swap to apply must have helped
- ok I get na's - I think that was the cause of the original error

```{r}
run_miss_methyl(biomart_genes, test_anno)
```

```{r}
test_b
```



