---
title: "04_compareDM"
author: "Caitlin Page"
date: "2025-02-21"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: inline
---

## Introduction
Comparison of differential methylation results.


## Packages

```{r}
library(plyranges)
library(dplyr)
library(ggplot2)
library(ggVennDiagram)
library(ComplexUpset)
```


## Pre-processed files
```{r}
wgbs_counts <- readRDS("../output/wgbs_counts.rds")
#sequencing
dmrcate_seq_anno <- readRDS("../output/dmrcate_seq_anno.rds")
dmrcate_seq_anno <- dmrcate_seq_anno@ranges %>% data.frame() %>%
  mutate(pos = paste0(seqnames, "-", start))
dmrcate_seq_anno <- dmrcate_seq_anno %>% .[order(.$ind.fdr),] %>% mutate(rank = 1:n())

dmrcate_seq_dmr <- readRDS("../output/dmrcate_seq_dmr.rds")
dmrcate_seq_dmr <- dmrcate_seq_dmr %>% data.frame()

bsseq_dmrs <- readRDS("../output/bsseq_dmrs.rds") 
bsseq_res <- readRDS("../output/bsseq_res.rds")

seq_edge_res <- readRDS("../output/seq_edge_res.rds")

#array
DMPs <- readRDS("../output/array_dmps.rds") 
DMPs$pos <- paste0(DMPs$seqnames, "-", DMPs$start)

dmrcate_array_dmr <- readRDS("../output/dmrcate_array_dmr.rds")
dmrcate_array_dmr <- dmrcate_array_dmr %>% data.frame()

dmrcate_array_anno <- readRDS("../output/dmrcate_array_anno.rds")
dmrcate_array_anno <- dmrcate_array_anno@ranges %>% data.frame() %>%
  mutate(pos = paste0(seqnames, "-", start)) %>% .[order(.$ind.fdr),] %>% mutate(rank = 1:n())
#dmrcate_array_anno <- dmrcate_array_anno %>% mutate(cg_id = anno_array[match(.$pos, anno_array$position), "Name"])
```


## Filtering parameters

# minfi
```{r}
head(DMPs)
```
```{r}
nrow(DMPs)
```
* recommended filtering parameter for GST is maximum of 10k CpGs.
* using more results in more noise - because you end up identifying every gene as significant.

# dmrcate array
```{r}
head(dmrcate_array_dmr)
```

```{r}
dmrcate_array_dmr_filt <- dmrcate_array_dmr %>% filter(no.cpgs > 3, abs(meandiff) > 0.1)
c(array = nrow(dmrcate_array_dmr), filt = nrow(dmrcate_array_dmr_filt))
```

Sum of CpGs across those DMRs
```{r}
c(array = sum(dmrcate_array_dmr$no.cpgs), filt = sum(dmrcate_array_dmr$no.cpgs))
```

# bsseq
```{r}
head(bsseq_dmrs)
```

```{r}
bsseq_dmrs_filt <- bsseq_dmrs %>% filter(abs(areaStat) > 100)
c(dmrs = nrow(bsseq_dmrs), filt = nrow(bsseq_dmrs_filt))
```

Sum of CpGs across DMRs
```{r}
c(dmrs = sum(bsseq_dmrs$n), filt = sum(bsseq_dmrs_filt$n))
```


# dmrcate seq
```{r}
head(dmrcate_seq_dmr)
```

```{r}
c(dmrs = nrow(dmrcate_seq_dmr), `5k` = 5000, `1k` = 1000)
```
```{r}
c(dmrs = sum(dmrcate_seq_dmr$no.cpgs), `5k` = sum(dmrcate_seq_dmr[1:5000,"no.cpgs"]), 
  `1k` = sum(dmrcate_seq_dmr[1:1000,"no.cpgs"]))
```


# edgeR
```{r}
head(seq_edge_res)
```



## compare dm methods

```{r}
compare_dm <- wgbs_counts[,1:4]
```

```{r}
overlap_dm <- find_overlaps(as_granges(dmrcate_seq_dmr[1:5000,]), as_granges(compare_dm)) %>% data.frame()
compare_dm <- compare_dm %>% mutate(seq_dmrcate_dmr_5k = ifelse(pos %in% overlap_dm$pos, TRUE, FALSE))
overlap_dm <- find_overlaps(as_granges(bsseq_dmrs_filt), as_granges(compare_dm)) %>% data.frame()
compare_dm <- compare_dm %>% mutate(bsseq_dmr_area100 = ifelse(pos %in% overlap_dm$pos, TRUE, FALSE))

compare_dm <- compare_dm %>% mutate(seq_edger_pos = seq_edge_res[match(.$pos, rownames(seq_edge_res)), "signif.05"])

overlap_dm <- find_overlaps(as_granges(dmrcate_array_dmr_filt), as_granges(compare_dm)) %>% data.frame()
compare_dm <- compare_dm %>% mutate(array_dmrcate_filt = ifelse(pos %in% overlap_dm$pos, TRUE, FALSE),
                                    array_minfi_filt = ifelse(pos %in% DMPs[1:10000,]$pos, TRUE, FALSE))
```


```{r}
upset(filter(compare_dm, seq_dmrcate_dmr_5k != FALSE| bsseq_dmr_area100 != FALSE| seq_edger_pos != FALSE| array_dmrcate_filt != FALSE| array_minfi_filt != FALSE), intersect = colnames(compare_dm)[5:9])
```
```{r}
rbind(cbind(width = dmrcate_seq_dmr$width, method = "dmrcate"),
      cbind(width = dmrcate_seq_dmr[1:5000,]$width, method = "dmrcate_5k"),
      cbind(width = dmrcate_seq_dmr[1:1000,]$width, method = "dmrcate_1k"),
      cbind(width = bsseq_dmrs_filt$width, method = "bsseq_area100")) %>% data.frame() %>% mutate(width = as.double(width)) %>%
  ggplot(aes(x = width, colour = method)) +
  geom_density()

rbind(cbind(width = dmrcate_seq_dmr$width, method = "dmrcate"),
      cbind(width = dmrcate_seq_dmr[1:5000,]$width, method = "dmrcate_5k"),
      cbind(width = dmrcate_seq_dmr[1:1000,]$width, method = "dmrcate_1k"),
      cbind(width = bsseq_dmrs_filt$width, method = "bsseq_area100")) %>% data.frame() %>% mutate(width = as.double(width)) %>%
  ggplot(aes(x = log2(width), colour = method)) +
  geom_density()
```

```{r}
rbind(cbind(no.cpgs = dmrcate_seq_dmr$no.cpgs, method = "dmrcate"),
      cbind(no.cpgs = dmrcate_seq_dmr[1:5000,]$no.cpgs, method = "dmrcate_5k"),
      cbind(no.cpgs = dmrcate_seq_dmr[1:1000,]$no.cpgs, method = "dmrcate_1k"),
      cbind(no.cpgs = bsseq_dmrs_filt$n, method = "bsseq_area100")) %>% data.frame() %>% mutate(no.cpgs = as.double(no.cpgs)) %>%
  ggplot(aes(x = no.cpgs, colour = method)) +
  geom_density()

rbind(cbind(no.cpgs = dmrcate_seq_dmr$no.cpgs, method = "dmrcate"),
      cbind(no.cpgs = dmrcate_seq_dmr[1:5000,]$no.cpgs, method = "dmrcate_5k"),
      cbind(no.cpgs = dmrcate_seq_dmr[1:1000,]$no.cpgs, method = "dmrcate_1k"),
      cbind(no.cpgs = bsseq_dmrs_filt$n, method = "bsseq_area100")) %>% data.frame() %>% mutate(no.cpgs = as.double(no.cpgs)) %>%
  ggplot(aes(x = log2(no.cpgs), colour = method)) +
  geom_density()
```



```{r}
upset(compare_dm[1:100000,], intersect = colnames(compare_dm)[5:9])
```



* ditch the overlapping DMRs for a minute because the venn diagrams are always ugly and they get messy

* upset plot
* and we can look at all cpgs - and is that cpg in a dmr for that method?
** this probably won't help with comparing to microarray - as diff starting cpgs
** but could shove it in anyway - or could just run on overlap

```{r}
cg_pos <- wgbs_counts[,c("pos", "seqnames", "start")]
cg_pos$end <- cg_pos$start
```
```{r}
cg_in_dmr <- cg_pos
```
```{r}
overlap <- find_overlaps(as_granges(cg_in_dmr), as_granges(bsseq_dmrs_filt)) %>% data.frame()
cg_in_dmr$bsseq <- ifelse(cg_in_dmr$pos %in% overlap$pos, TRUE, FALSE)

overlap <- find_overlaps(as_granges(cg_in_dmr), as_granges(dmrcate_seq_dmr)) %>% data.frame()
cg_in_dmr$dmrcate <- ifelse(cg_in_dmr$pos %in% overlap$pos, TRUE, FALSE)

overlap <- find_overlaps(as_granges(cg_in_dmr), as_granges(dmrcate_array_dmr)) %>% data.frame()
cg_in_dmr$dmrcate_array <- ifelse(cg_in_dmr$pos %in% overlap$pos, TRUE, FALSE)
```
```{r}
ggVennDiagram(list(bsseq = filter(cg_in_dmr, bsseq == TRUE)$pos, dmrcate = filter(cg_in_dmr, dmrcate == TRUE)$pos))
ggVennDiagram(list(bsseq = filter(cg_in_dmr, bsseq != TRUE)$pos, dmrcate = filter(cg_in_dmr, dmrcate != TRUE)$pos))
```
* agree best about what is not significant
* most of dmrcate significant is also identified by bsseq
** this was higher when using the unfiltered bsseq dmrs 

```{r}
ggVennDiagram(list(bsseq = filter(cg_in_dmr, bsseq == TRUE)$pos, dmrcate = filter(cg_in_dmr, dmrcate == TRUE)$pos, 
                   array_dmr = filter(cg_in_dmr, dmrcate_array == TRUE)$pos))
ggVennDiagram(list(bsseq = filter(cg_in_dmr, bsseq != TRUE)$pos, dmrcate = filter(cg_in_dmr, dmrcate != TRUE)$pos, 
                   array_dmr = filter(cg_in_dmr, dmrcate_array != TRUE)$pos))
```
* this doesn't acknowledge that the array does not cover all the sites seq does - it would just add to the FALSE




##################################################

```{r}
dmrcate_overlap_pos <- find_overlaps(as_granges(dmrcate_seq_anno), as_granges(dmrcate_array_anno), maxgap = 2) %>% data.frame()
ggVennDiagram(list(array_rank = dmrcate_array_anno$rank, overlap = unique(dmrcate_overlap_pos$rank.y)))
ggVennDiagram(list(seq_rank = dmrcate_seq_anno$rank, overlap = unique(dmrcate_overlap_pos$rank.x)))

# just the significant
dmrcate_overlap_pos <- find_overlaps(as_granges(filter(dmrcate_seq_anno, is.sig == TRUE)), as_granges(filter(dmrcate_array_anno, is.sig == TRUE)), maxgap = 2) %>% data.frame()
ggVennDiagram(list(array_rank = filter(dmrcate_array_anno, is.sig == TRUE)$rank, overlap = unique(dmrcate_overlap_pos$rank.y)))
ggVennDiagram(list(seq_rank = filter(dmrcate_seq_anno, is.sig == TRUE)$rank, overlap = unique(dmrcate_overlap_pos$rank.x)))
```



```{r}
dmrcate_seq_dmr$num <- 1:nrow(dmrcate_seq_dmr)
```

```{r}
dmrcate_overlap_dmr <- find_overlaps(as_granges(dmrcate_seq_dmr), as_granges(dmrcate_array_dmr)) %>% data.frame()
ggVennDiagram(list(array_rank = dmrcate_array_dmr$rank, overlap = unique(dmrcate_overlap_dmr$rank)))
ggVennDiagram(list(seq_rank = dmrcate_seq_dmr$num, overlap = unique(dmrcate_overlap_dmr$num)))
```

* same method - still very little results overlap

* I need to find myself a matched dataset


#########

```{r}
overlap_dmr <- find_overlaps(as_granges(dmrcate_seq_dmr), as_granges(bsseq_dmrs)) %>% data.frame()
ggVennDiagram(list(bsseq_rank = bsseq_dmrs$bsseq_rank, overlap = unique(overlap_dmr$bsseq_rank)))
ggVennDiagram(list(seq_rank = dmrcate_seq_dmr$num, overlap = unique(dmrcate_overlap_dmr$num)))
```

