---
title: "Home"
site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    toc: false
editor_options:
  chunk_output_type: console
---

Welcome to my research website.

The main analyses conducted uses B cell type and NK cell type samples, 
under the hg19 genome annotation.
* Microarray: EPICv1 platform. [GSE110554](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE110554)
* WGBS [GSE186458](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE186458)

Analysis:

1. Differential methylation methods and comparison

  a) [Microarray](00_differentialMethylationArray.html)
  
  b) Sequencing *tbd 00_differentialMethylationSequencing
  
  c) Compare microarray and sequencing DMRcate results 

2. [Methylation microarray gene set testing analysis](00_microarrayOntologyAnalysis.Rmd)

3. Identify bias for correction in gene set testing

  a) [Explore CpG distribution across genome bins](01_exploreBiasByBins.Rmd)
  
4. Develop gene set testing method for methylation sequencing

5. Validate gene set testing method for methylation sequencing

6. Applying gene set testing method to different data sets

  a) Long read Analysis 00_longRead

```{r}
	
#save.image(file='myEnvironment.RData')
```

https://www.techcoil.com/blog/how-to-save-and-load-environment-objects-in-r/#google_vignette
```{r}
#load('myEnvironment.RData')
```





[analysis/00_exploreCpGDistribution]
* plots comparing array and sequencing
** distance between cpgs, annotation to cpg and gene features



