---
title: "Home"
site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    toc: false
editor_options:
  chunk_output_type: console
---

Welcome to my research website.

The main analyses conducted uses B cell type and NK cell type samples, 
under the hg19 genome annotation.
* Microarray: EPICv1 platform. [GSE110554](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE110554)
* WGBS [GSE186458](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE186458)


* Initial analysis and development of GST methods was done on results from bsseq.
* This was the method that ran and worked well.
* When revisiting analysis - retried DMRcate and it ran (didn't initially).
* It was attractive as it had a method for array and seq, and found less DMRs (bsseq was very noisy)
* original microarray analysis was done with minfi - and then repeated with DMRcate
* results from both methods will be included in the various steps - as the original development, and as the subsequent analysis.


Analysis:
0. [Summary of decisions]
0. [Datasets used in the analyses](00_datasets.html) 

1. [Exploration of CpG sites](00_exploreCpGDistrib.html)

1. Differential methylation methods and comparison

  a) [Microarray](00_differentialMethylationArray.html)
  
  b) [Sequencing](00_differentialMethylationSequencing.html)
  
  [Compare Array and Sequencing results](04_compareDM.html)
  
  [Compare distributions of tested array and sequencing sites](02_cpgAvailVsTested.html)
  
  c) Compare CpG level and DMR results
  
  c) Compare microarray and sequencing DMRcate results 

2. Annotating DMRs to genes

  a) [Collating genes](00_obtainGenes.html)

  a) Comparing significant genes

  a) Comparing microarray and sequencing genes

2. [Methylation microarray gene set testing analysis](00_microarrayOntologyAnalysis.html)

3. Identify bias for correction in gene set testing

  a) [Explore CpG distribution across genome bins](01_exploreBiasByBins.html)
  
  b) [Explore gene based biases](02_exploreBiasGenes.html) 

4. Gene Set Testing
  [Summary page]
  a) [Explore gene sets, ways to obtain, relationships, and bias](00_exploreCpGGST.html)
  b) [Converting ]

  b) [Reviewing code of existing methods](03_howToGST.html)
  
  [Develop gene set testing method for methylation sequencing](03a_developGeneSetTest.html)
  
4. [Evaluate gene set testing method for methylation sequencing](03_runGST.html)

[Compare array and sequencing gene set testing results]

6. Applying gene set testing method to different data sets

  a) [Long read Analysis](00_longReadAnalysis.html)
  
  b) [Monocyte vs Granulocyte](07_monoGranAnalysis.html)
  
  b) [cd4 vs cd8](07_tCellAnalysis.html)
  
  b) [Broad Gene Set Database]
  
7. [Final Result Plots]
```{r}
	
save.image(file='myEnvironment.RData')
```

https://www.techcoil.com/blog/how-to-save-and-load-environment-objects-in-r/#google_vignette
```{r}
#load('myEnvironment.RData')
```




